{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "answering-ranch",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import pickle\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.spatial.distance as ssd\n",
    "import matplotlib as mpl\n",
    "from matplotlib.colors import LinearSegmentedColormap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "nominated-exchange",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'> is in the pickle\n",
      "(666, 61) total bins\n",
      "656 bins after removing controls\n",
      "61 columns\n",
      "63 columns after adding max/thresholded abund cols\n"
     ]
    }
   ],
   "source": [
    "# read in the qualtax dataframe to get taxonomy and completeness of bins\n",
    "with open( '/Volumes/KeithSSD/CBFunctions/data/outputs/qual_tax_full.p', \"rb\" ) as fh:\n",
    "    qualtax = pickle.load(fh)\n",
    "\n",
    "print(type(qualtax), 'is in the pickle')\n",
    "\n",
    "# formatting\n",
    "conversions = [('Contamination', 2), ('Completeness', 2), ('red_value', 3)]\n",
    "for i, j in conversions:\n",
    "    qualtax[i] = qualtax[i].astype(float)\n",
    "    ibool = qualtax.Contamination.notnull()\n",
    "    qualtax.loc[ibool, i] = qualtax.loc[ibool, i].round(j)\n",
    "\n",
    "print(qualtax.shape, 'total bins')\n",
    "\n",
    "# remove controls\n",
    "qualtax = qualtax.loc[[i for i in qualtax.index if not i.startswith('Zymo')], :].copy()\n",
    "\n",
    "print(qualtax.shape[0], 'bins after removing controls')\n",
    "print(qualtax.shape[1], 'columns')\n",
    "\n",
    "with open('/Volumes/KeithSSD/CBFunctions/data/outputs/bin_abund_n_metadata.p', \"rb\" ) as fh:\n",
    "    outpkg = pickle.load(fh)\n",
    "\n",
    "name_correction = outpkg['renamer']\n",
    "metadata = outpkg['meta']\n",
    "cb33_abnd, cbTr_abnd = outpkg['cb33_abnd'], outpkg['cbTr_abnd']\n",
    "\n",
    "sig_pres_fxn = lambda x: (x > 1).sum()\n",
    "abund_srs = cbTr_abnd.apply(sig_pres_fxn).append(cb33_abnd.apply(sig_pres_fxn))\n",
    "abund_srs = (abund_srs).round(1)\n",
    "abund_srs.name = '% samples w/ Mean Cov > 2x'\n",
    "\n",
    "abund_srs2 = cbTr_abnd.apply(np.max).append(cb33_abnd.apply(np.max))\n",
    "abund_srs2 = (abund_srs2).round(1)\n",
    "abund_srs2.name = 'MaxCopyNumber'\n",
    "\n",
    "qualtax = qualtax.join(abund_srs).join(abund_srs2)\n",
    "print(qualtax.shape[1], 'columns after adding max/thresholded abund cols')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "floral-lexington",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "52 completeish bins\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Completeness</th>\n",
       "      <th>Contamination</th>\n",
       "      <th>red_value</th>\n",
       "      <th>aa_percent</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>% samples w/ Mean Cov &gt; 2x</th>\n",
       "      <th>MaxCopyNumber</th>\n",
       "      <th>N50 (contigs)</th>\n",
       "      <th>GenomeSizeMbp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.375</th>\n",
       "      <td>88.39</td>\n",
       "      <td>9.98</td>\n",
       "      <td>0.687</td>\n",
       "      <td>92.94</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiohalospirales</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>14.5</td>\n",
       "      <td>17860</td>\n",
       "      <td>1.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.104</th>\n",
       "      <td>97.09</td>\n",
       "      <td>29.45</td>\n",
       "      <td>0.991</td>\n",
       "      <td>85.48</td>\n",
       "      <td>Crenarchaeota</td>\n",
       "      <td>Nitrososphaeria</td>\n",
       "      <td>Nitrososphaerales</td>\n",
       "      <td>Nitrosopumilaceae</td>\n",
       "      <td>Nitrosopumilus</td>\n",
       "      <td></td>\n",
       "      <td>6</td>\n",
       "      <td>7.1</td>\n",
       "      <td>21735</td>\n",
       "      <td>1.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.42</th>\n",
       "      <td>81.03</td>\n",
       "      <td>6.46</td>\n",
       "      <td>NaN</td>\n",
       "      <td>81.88</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Flavobacteriales</td>\n",
       "      <td>Cryomorphaceae</td>\n",
       "      <td>TMED14</td>\n",
       "      <td>TMED14 sp002381225</td>\n",
       "      <td>3</td>\n",
       "      <td>6.3</td>\n",
       "      <td>17268</td>\n",
       "      <td>1.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.182</th>\n",
       "      <td>90.11</td>\n",
       "      <td>9.89</td>\n",
       "      <td>0.712</td>\n",
       "      <td>91.73</td>\n",
       "      <td>Marinisomatota</td>\n",
       "      <td>UBA8477</td>\n",
       "      <td>UBA8477</td>\n",
       "      <td>UBA8477</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5</td>\n",
       "      <td>6.1</td>\n",
       "      <td>85486</td>\n",
       "      <td>4.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.318</th>\n",
       "      <td>100.00</td>\n",
       "      <td>73.28</td>\n",
       "      <td>0.980</td>\n",
       "      <td>93.99</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Nanopelagicales</td>\n",
       "      <td>Nanopelagicaceae</td>\n",
       "      <td>MAG-120802</td>\n",
       "      <td></td>\n",
       "      <td>11</td>\n",
       "      <td>5.6</td>\n",
       "      <td>12215</td>\n",
       "      <td>1.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.2</th>\n",
       "      <td>93.10</td>\n",
       "      <td>15.20</td>\n",
       "      <td>0.919</td>\n",
       "      <td>88.10</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Nanopelagicales</td>\n",
       "      <td>Nanopelagicaceae</td>\n",
       "      <td>MAG-120802</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>5.5</td>\n",
       "      <td>19441</td>\n",
       "      <td>1.77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.5</th>\n",
       "      <td>83.16</td>\n",
       "      <td>8.93</td>\n",
       "      <td>0.856</td>\n",
       "      <td>84.07</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Acidimicrobiia</td>\n",
       "      <td>Microtrichales</td>\n",
       "      <td>Ilumatobacteraceae</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>6</td>\n",
       "      <td>5.5</td>\n",
       "      <td>16313</td>\n",
       "      <td>1.66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.263</th>\n",
       "      <td>90.14</td>\n",
       "      <td>17.11</td>\n",
       "      <td>0.818</td>\n",
       "      <td>88.27</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiomicrospirales</td>\n",
       "      <td>Thioglobaceae</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>5.4</td>\n",
       "      <td>10418</td>\n",
       "      <td>1.87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.14</th>\n",
       "      <td>82.05</td>\n",
       "      <td>7.26</td>\n",
       "      <td>0.970</td>\n",
       "      <td>81.69</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Acidimicrobiia</td>\n",
       "      <td>Microtrichales</td>\n",
       "      <td>Ilumatobacteraceae</td>\n",
       "      <td>Casp-actino8</td>\n",
       "      <td></td>\n",
       "      <td>11</td>\n",
       "      <td>5.2</td>\n",
       "      <td>22943</td>\n",
       "      <td>2.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.233</th>\n",
       "      <td>93.79</td>\n",
       "      <td>34.78</td>\n",
       "      <td>0.885</td>\n",
       "      <td>91.51</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Chromatiales</td>\n",
       "      <td>Sedimenticolaceae</td>\n",
       "      <td>MONJU</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4151</td>\n",
       "      <td>2.28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.12</th>\n",
       "      <td>81.71</td>\n",
       "      <td>40.55</td>\n",
       "      <td>0.911</td>\n",
       "      <td>70.42</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Pelagibacterales</td>\n",
       "      <td>Pelagibacteraceae</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>26</td>\n",
       "      <td>4.8</td>\n",
       "      <td>17867</td>\n",
       "      <td>1.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.149</th>\n",
       "      <td>97.90</td>\n",
       "      <td>95.24</td>\n",
       "      <td>0.688</td>\n",
       "      <td>93.08</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiohalospirales</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>6</td>\n",
       "      <td>4.4</td>\n",
       "      <td>15544</td>\n",
       "      <td>3.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.372</th>\n",
       "      <td>84.81</td>\n",
       "      <td>7.24</td>\n",
       "      <td>0.919</td>\n",
       "      <td>90.65</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Nanopelagicales</td>\n",
       "      <td>Nanopelagicaceae</td>\n",
       "      <td>MAG-120802</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>4.3</td>\n",
       "      <td>32513</td>\n",
       "      <td>1.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.236</th>\n",
       "      <td>95.35</td>\n",
       "      <td>7.26</td>\n",
       "      <td>0.931</td>\n",
       "      <td>89.09</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Chromatiales</td>\n",
       "      <td>Sedimenticolaceae</td>\n",
       "      <td>Sedimenticola</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>4.3</td>\n",
       "      <td>21011</td>\n",
       "      <td>3.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.362</th>\n",
       "      <td>87.03</td>\n",
       "      <td>7.20</td>\n",
       "      <td>NaN</td>\n",
       "      <td>89.46</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Nanopelagicales</td>\n",
       "      <td>Nanopelagicaceae</td>\n",
       "      <td>Planktophila</td>\n",
       "      <td>Planktophila sp001438925</td>\n",
       "      <td>2</td>\n",
       "      <td>2.9</td>\n",
       "      <td>12225</td>\n",
       "      <td>1.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.191</th>\n",
       "      <td>88.97</td>\n",
       "      <td>2.97</td>\n",
       "      <td>0.819</td>\n",
       "      <td>89.52</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiomicrospirales</td>\n",
       "      <td>Thioglobaceae</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>2.8</td>\n",
       "      <td>8445</td>\n",
       "      <td>1.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.196</th>\n",
       "      <td>92.48</td>\n",
       "      <td>12.41</td>\n",
       "      <td>0.820</td>\n",
       "      <td>90.08</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>GCA-001735895</td>\n",
       "      <td>GCA-001735895</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>2.7</td>\n",
       "      <td>24647</td>\n",
       "      <td>5.29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.111</th>\n",
       "      <td>84.74</td>\n",
       "      <td>18.58</td>\n",
       "      <td>NaN</td>\n",
       "      <td>72.22</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteriia</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>Cyanobiaceae</td>\n",
       "      <td>Vulcanococcus</td>\n",
       "      <td>Vulcanococcus sp000179255</td>\n",
       "      <td>3</td>\n",
       "      <td>2.6</td>\n",
       "      <td>5654</td>\n",
       "      <td>2.24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.167</th>\n",
       "      <td>84.99</td>\n",
       "      <td>12.74</td>\n",
       "      <td>0.523</td>\n",
       "      <td>88.87</td>\n",
       "      <td>Myxococcota</td>\n",
       "      <td>UBA6777</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>2.6</td>\n",
       "      <td>10422</td>\n",
       "      <td>6.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.174</th>\n",
       "      <td>85.92</td>\n",
       "      <td>27.07</td>\n",
       "      <td>0.940</td>\n",
       "      <td>85.22</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rhodobacterales</td>\n",
       "      <td>Rhodobacteraceae</td>\n",
       "      <td>Marivivens</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>2.5</td>\n",
       "      <td>13294</td>\n",
       "      <td>3.41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.156</th>\n",
       "      <td>93.41</td>\n",
       "      <td>3.66</td>\n",
       "      <td>0.769</td>\n",
       "      <td>92.08</td>\n",
       "      <td>Krumholzibacteriota</td>\n",
       "      <td>Krumholzibacteria</td>\n",
       "      <td>LZORAL124-64-63</td>\n",
       "      <td>LZORAL124-64-63</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>2.4</td>\n",
       "      <td>17949</td>\n",
       "      <td>3.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.407</th>\n",
       "      <td>91.45</td>\n",
       "      <td>12.82</td>\n",
       "      <td>0.855</td>\n",
       "      <td>94.37</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Acidimicrobiia</td>\n",
       "      <td>Microtrichales</td>\n",
       "      <td>Ilumatobacteraceae</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>2.3</td>\n",
       "      <td>20425</td>\n",
       "      <td>2.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.101</th>\n",
       "      <td>93.10</td>\n",
       "      <td>19.28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>88.67</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Acidimicrobiia</td>\n",
       "      <td>Microtrichales</td>\n",
       "      <td>Ilumatobacteraceae</td>\n",
       "      <td>UBA3006</td>\n",
       "      <td>UBA3006 sp001438985</td>\n",
       "      <td>5</td>\n",
       "      <td>2.3</td>\n",
       "      <td>9393</td>\n",
       "      <td>2.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.212</th>\n",
       "      <td>92.40</td>\n",
       "      <td>5.73</td>\n",
       "      <td>0.704</td>\n",
       "      <td>86.75</td>\n",
       "      <td>Verrucomicrobiota</td>\n",
       "      <td>Kiritimatiellae</td>\n",
       "      <td>UBA8416</td>\n",
       "      <td>UBA8416</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>2.3</td>\n",
       "      <td>12322</td>\n",
       "      <td>4.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.45</th>\n",
       "      <td>88.79</td>\n",
       "      <td>25.24</td>\n",
       "      <td>0.900</td>\n",
       "      <td>84.13</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Acidimicrobiia</td>\n",
       "      <td>Microtrichales</td>\n",
       "      <td>Ilumatobacteraceae</td>\n",
       "      <td>Casp-actino5</td>\n",
       "      <td></td>\n",
       "      <td>7</td>\n",
       "      <td>2.2</td>\n",
       "      <td>10935</td>\n",
       "      <td>2.13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.395</th>\n",
       "      <td>85.05</td>\n",
       "      <td>1.80</td>\n",
       "      <td>0.931</td>\n",
       "      <td>82.94</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Chromatiales</td>\n",
       "      <td>Sedimenticolaceae</td>\n",
       "      <td>Sedimenticola</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>2.2</td>\n",
       "      <td>52182</td>\n",
       "      <td>3.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.40</th>\n",
       "      <td>83.53</td>\n",
       "      <td>3.67</td>\n",
       "      <td>0.969</td>\n",
       "      <td>79.80</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Flavobacteriales</td>\n",
       "      <td>Flavobacteriaceae</td>\n",
       "      <td>MS024-2A</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>2.2</td>\n",
       "      <td>86578</td>\n",
       "      <td>1.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.368</th>\n",
       "      <td>97.23</td>\n",
       "      <td>26.67</td>\n",
       "      <td>0.884</td>\n",
       "      <td>97.30</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Chromatiales</td>\n",
       "      <td>Sedimenticolaceae</td>\n",
       "      <td>MONJU</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>2.1</td>\n",
       "      <td>14377</td>\n",
       "      <td>2.65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.358</th>\n",
       "      <td>91.38</td>\n",
       "      <td>9.72</td>\n",
       "      <td>0.940</td>\n",
       "      <td>94.84</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rhodobacterales</td>\n",
       "      <td>Rhodobacteraceae</td>\n",
       "      <td>Marivivens</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>2.1</td>\n",
       "      <td>57260</td>\n",
       "      <td>3.56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.246</th>\n",
       "      <td>89.20</td>\n",
       "      <td>13.59</td>\n",
       "      <td>0.989</td>\n",
       "      <td>76.83</td>\n",
       "      <td>Crenarchaeota</td>\n",
       "      <td>Nitrososphaeria</td>\n",
       "      <td>Nitrososphaerales</td>\n",
       "      <td>Nitrosopumilaceae</td>\n",
       "      <td>Nitrosopumilus</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>13630</td>\n",
       "      <td>1.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.63</th>\n",
       "      <td>91.99</td>\n",
       "      <td>13.27</td>\n",
       "      <td>0.951</td>\n",
       "      <td>82.50</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteriia</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>Cyanobiaceae</td>\n",
       "      <td>Vulcanococcus</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6759</td>\n",
       "      <td>2.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.417</th>\n",
       "      <td>91.88</td>\n",
       "      <td>34.42</td>\n",
       "      <td>0.976</td>\n",
       "      <td>92.70</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Acidimicrobiia</td>\n",
       "      <td>Microtrichales</td>\n",
       "      <td>Ilumatobacteraceae</td>\n",
       "      <td>BACL27</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>1.9</td>\n",
       "      <td>8162</td>\n",
       "      <td>2.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.12</th>\n",
       "      <td>87.31</td>\n",
       "      <td>80.22</td>\n",
       "      <td>0.887</td>\n",
       "      <td>78.69</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Pelagibacterales</td>\n",
       "      <td>Pelagibacteraceae</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>1.8</td>\n",
       "      <td>5189</td>\n",
       "      <td>1.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.188</th>\n",
       "      <td>98.12</td>\n",
       "      <td>19.89</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.37</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Flavobacteriales</td>\n",
       "      <td>Cryomorphaceae</td>\n",
       "      <td>TMED14</td>\n",
       "      <td>TMED14 sp002381225</td>\n",
       "      <td>2</td>\n",
       "      <td>1.8</td>\n",
       "      <td>36065</td>\n",
       "      <td>2.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.408</th>\n",
       "      <td>99.34</td>\n",
       "      <td>51.18</td>\n",
       "      <td>0.905</td>\n",
       "      <td>93.73</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteriia</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>Cyanobiaceae</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>1.7</td>\n",
       "      <td>30189</td>\n",
       "      <td>3.33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Completeness  Contamination  red_value  aa_percent  \\\n",
       "CBrest.bin.375         88.39           9.98      0.687       92.94   \n",
       "CB33.bin.104           97.09          29.45      0.991       85.48   \n",
       "CB33.bin.42            81.03           6.46        NaN       81.88   \n",
       "CB33.bin.182           90.11           9.89      0.712       91.73   \n",
       "CBrest.bin.318        100.00          73.28      0.980       93.99   \n",
       "CB33.bin.2             93.10          15.20      0.919       88.10   \n",
       "CB33.bin.5             83.16           8.93      0.856       84.07   \n",
       "CBrest.bin.263         90.14          17.11      0.818       88.27   \n",
       "CB33.bin.14            82.05           7.26      0.970       81.69   \n",
       "CB33.bin.233           93.79          34.78      0.885       91.51   \n",
       "CBrest.bin.12          81.71          40.55      0.911       70.42   \n",
       "CB33.bin.149           97.90          95.24      0.688       93.08   \n",
       "CBrest.bin.372         84.81           7.24      0.919       90.65   \n",
       "CB33.bin.236           95.35           7.26      0.931       89.09   \n",
       "CBrest.bin.362         87.03           7.20        NaN       89.46   \n",
       "CB33.bin.191           88.97           2.97      0.819       89.52   \n",
       "CB33.bin.196           92.48          12.41      0.820       90.08   \n",
       "CB33.bin.111           84.74          18.58        NaN       72.22   \n",
       "CB33.bin.167           84.99          12.74      0.523       88.87   \n",
       "CB33.bin.174           85.92          27.07      0.940       85.22   \n",
       "CB33.bin.156           93.41           3.66      0.769       92.08   \n",
       "CBrest.bin.407         91.45          12.82      0.855       94.37   \n",
       "CB33.bin.101           93.10          19.28        NaN       88.67   \n",
       "CB33.bin.212           92.40           5.73      0.704       86.75   \n",
       "CB33.bin.45            88.79          25.24      0.900       84.13   \n",
       "CBrest.bin.395         85.05           1.80      0.931       82.94   \n",
       "CBrest.bin.40          83.53           3.67      0.969       79.80   \n",
       "CBrest.bin.368         97.23          26.67      0.884       97.30   \n",
       "CBrest.bin.358         91.38           9.72      0.940       94.84   \n",
       "CBrest.bin.246         89.20          13.59      0.989       76.83   \n",
       "CB33.bin.63            91.99          13.27      0.951       82.50   \n",
       "CBrest.bin.417         91.88          34.42      0.976       92.70   \n",
       "CB33.bin.12            87.31          80.22      0.887       78.69   \n",
       "CBrest.bin.188         98.12          19.89        NaN       98.37   \n",
       "CBrest.bin.408         99.34          51.18      0.905       93.73   \n",
       "\n",
       "                             Phylum                Class              Order  \\\n",
       "CBrest.bin.375       Proteobacteria  Gammaproteobacteria   Thiohalospirales   \n",
       "CB33.bin.104          Crenarchaeota      Nitrososphaeria  Nitrososphaerales   \n",
       "CB33.bin.42            Bacteroidota          Bacteroidia   Flavobacteriales   \n",
       "CB33.bin.182         Marinisomatota              UBA8477            UBA8477   \n",
       "CBrest.bin.318     Actinobacteriota       Actinobacteria    Nanopelagicales   \n",
       "CB33.bin.2         Actinobacteriota       Actinobacteria    Nanopelagicales   \n",
       "CB33.bin.5         Actinobacteriota       Acidimicrobiia     Microtrichales   \n",
       "CBrest.bin.263       Proteobacteria  Gammaproteobacteria  Thiomicrospirales   \n",
       "CB33.bin.14        Actinobacteriota       Acidimicrobiia     Microtrichales   \n",
       "CB33.bin.233         Proteobacteria  Gammaproteobacteria       Chromatiales   \n",
       "CBrest.bin.12        Proteobacteria  Alphaproteobacteria   Pelagibacterales   \n",
       "CB33.bin.149         Proteobacteria  Gammaproteobacteria   Thiohalospirales   \n",
       "CBrest.bin.372     Actinobacteriota       Actinobacteria    Nanopelagicales   \n",
       "CB33.bin.236         Proteobacteria  Gammaproteobacteria       Chromatiales   \n",
       "CBrest.bin.362     Actinobacteriota       Actinobacteria    Nanopelagicales   \n",
       "CB33.bin.191         Proteobacteria  Gammaproteobacteria  Thiomicrospirales   \n",
       "CB33.bin.196         Proteobacteria  Gammaproteobacteria      GCA-001735895   \n",
       "CB33.bin.111          Cyanobacteria       Cyanobacteriia    Synechococcales   \n",
       "CB33.bin.167            Myxococcota              UBA6777                      \n",
       "CB33.bin.174         Proteobacteria  Alphaproteobacteria    Rhodobacterales   \n",
       "CB33.bin.156    Krumholzibacteriota    Krumholzibacteria    LZORAL124-64-63   \n",
       "CBrest.bin.407     Actinobacteriota       Acidimicrobiia     Microtrichales   \n",
       "CB33.bin.101       Actinobacteriota       Acidimicrobiia     Microtrichales   \n",
       "CB33.bin.212      Verrucomicrobiota      Kiritimatiellae            UBA8416   \n",
       "CB33.bin.45        Actinobacteriota       Acidimicrobiia     Microtrichales   \n",
       "CBrest.bin.395       Proteobacteria  Gammaproteobacteria       Chromatiales   \n",
       "CBrest.bin.40          Bacteroidota          Bacteroidia   Flavobacteriales   \n",
       "CBrest.bin.368       Proteobacteria  Gammaproteobacteria       Chromatiales   \n",
       "CBrest.bin.358       Proteobacteria  Alphaproteobacteria    Rhodobacterales   \n",
       "CBrest.bin.246        Crenarchaeota      Nitrososphaeria  Nitrososphaerales   \n",
       "CB33.bin.63           Cyanobacteria       Cyanobacteriia    Synechococcales   \n",
       "CBrest.bin.417     Actinobacteriota       Acidimicrobiia     Microtrichales   \n",
       "CB33.bin.12          Proteobacteria  Alphaproteobacteria   Pelagibacterales   \n",
       "CBrest.bin.188         Bacteroidota          Bacteroidia   Flavobacteriales   \n",
       "CBrest.bin.408        Cyanobacteria       Cyanobacteriia    Synechococcales   \n",
       "\n",
       "                            Family           Genus                    Species  \\\n",
       "CBrest.bin.375                                                                  \n",
       "CB33.bin.104     Nitrosopumilaceae  Nitrosopumilus                              \n",
       "CB33.bin.42         Cryomorphaceae          TMED14         TMED14 sp002381225   \n",
       "CB33.bin.182               UBA8477                                              \n",
       "CBrest.bin.318    Nanopelagicaceae      MAG-120802                              \n",
       "CB33.bin.2        Nanopelagicaceae      MAG-120802                              \n",
       "CB33.bin.5      Ilumatobacteraceae                                              \n",
       "CBrest.bin.263       Thioglobaceae                                              \n",
       "CB33.bin.14     Ilumatobacteraceae    Casp-actino8                              \n",
       "CB33.bin.233     Sedimenticolaceae           MONJU                              \n",
       "CBrest.bin.12    Pelagibacteraceae                                              \n",
       "CB33.bin.149                                                                    \n",
       "CBrest.bin.372    Nanopelagicaceae      MAG-120802                              \n",
       "CB33.bin.236     Sedimenticolaceae   Sedimenticola                              \n",
       "CBrest.bin.362    Nanopelagicaceae    Planktophila   Planktophila sp001438925   \n",
       "CB33.bin.191         Thioglobaceae                                              \n",
       "CB33.bin.196         GCA-001735895                                              \n",
       "CB33.bin.111          Cyanobiaceae   Vulcanococcus  Vulcanococcus sp000179255   \n",
       "CB33.bin.167                                                                    \n",
       "CB33.bin.174      Rhodobacteraceae      Marivivens                              \n",
       "CB33.bin.156       LZORAL124-64-63                                              \n",
       "CBrest.bin.407  Ilumatobacteraceae                                              \n",
       "CB33.bin.101    Ilumatobacteraceae         UBA3006        UBA3006 sp001438985   \n",
       "CB33.bin.212               UBA8416                                              \n",
       "CB33.bin.45     Ilumatobacteraceae    Casp-actino5                              \n",
       "CBrest.bin.395   Sedimenticolaceae   Sedimenticola                              \n",
       "CBrest.bin.40    Flavobacteriaceae        MS024-2A                              \n",
       "CBrest.bin.368   Sedimenticolaceae           MONJU                              \n",
       "CBrest.bin.358    Rhodobacteraceae      Marivivens                              \n",
       "CBrest.bin.246   Nitrosopumilaceae  Nitrosopumilus                              \n",
       "CB33.bin.63           Cyanobiaceae   Vulcanococcus                              \n",
       "CBrest.bin.417  Ilumatobacteraceae          BACL27                              \n",
       "CB33.bin.12      Pelagibacteraceae                                              \n",
       "CBrest.bin.188      Cryomorphaceae          TMED14         TMED14 sp002381225   \n",
       "CBrest.bin.408        Cyanobiaceae   Synechococcus                              \n",
       "\n",
       "                % samples w/ Mean Cov > 2x  MaxCopyNumber N50 (contigs)  \\\n",
       "CBrest.bin.375                           3           14.5         17860   \n",
       "CB33.bin.104                             6            7.1         21735   \n",
       "CB33.bin.42                              3            6.3         17268   \n",
       "CB33.bin.182                             5            6.1         85486   \n",
       "CBrest.bin.318                          11            5.6         12215   \n",
       "CB33.bin.2                               4            5.5         19441   \n",
       "CB33.bin.5                               6            5.5         16313   \n",
       "CBrest.bin.263                           2            5.4         10418   \n",
       "CB33.bin.14                             11            5.2         22943   \n",
       "CB33.bin.233                             4            5.0          4151   \n",
       "CBrest.bin.12                           26            4.8         17867   \n",
       "CB33.bin.149                             6            4.4         15544   \n",
       "CBrest.bin.372                           2            4.3         32513   \n",
       "CB33.bin.236                             3            4.3         21011   \n",
       "CBrest.bin.362                           2            2.9         12225   \n",
       "CB33.bin.191                             2            2.8          8445   \n",
       "CB33.bin.196                             2            2.7         24647   \n",
       "CB33.bin.111                             3            2.6          5654   \n",
       "CB33.bin.167                             3            2.6         10422   \n",
       "CB33.bin.174                             2            2.5         13294   \n",
       "CB33.bin.156                             3            2.4         17949   \n",
       "CBrest.bin.407                           3            2.3         20425   \n",
       "CB33.bin.101                             5            2.3          9393   \n",
       "CB33.bin.212                             3            2.3         12322   \n",
       "CB33.bin.45                              7            2.2         10935   \n",
       "CBrest.bin.395                           2            2.2         52182   \n",
       "CBrest.bin.40                            3            2.2         86578   \n",
       "CBrest.bin.368                           3            2.1         14377   \n",
       "CBrest.bin.358                           3            2.1         57260   \n",
       "CBrest.bin.246                           2            2.0         13630   \n",
       "CB33.bin.63                              2            2.0          6759   \n",
       "CBrest.bin.417                           2            1.9          8162   \n",
       "CB33.bin.12                              4            1.8          5189   \n",
       "CBrest.bin.188                           2            1.8         36065   \n",
       "CBrest.bin.408                           3            1.7         30189   \n",
       "\n",
       "                GenomeSizeMbp  \n",
       "CBrest.bin.375           1.56  \n",
       "CB33.bin.104             1.76  \n",
       "CB33.bin.42              1.72  \n",
       "CB33.bin.182             4.43  \n",
       "CBrest.bin.318           1.89  \n",
       "CB33.bin.2               1.77  \n",
       "CB33.bin.5               1.66  \n",
       "CBrest.bin.263           1.87  \n",
       "CB33.bin.14              2.06  \n",
       "CB33.bin.233             2.28  \n",
       "CBrest.bin.12            1.29  \n",
       "CB33.bin.149             3.09  \n",
       "CBrest.bin.372           1.72  \n",
       "CB33.bin.236             3.80  \n",
       "CBrest.bin.362           1.41  \n",
       "CB33.bin.191             1.26  \n",
       "CB33.bin.196             5.29  \n",
       "CB33.bin.111             2.24  \n",
       "CB33.bin.167             6.35  \n",
       "CB33.bin.174             3.41  \n",
       "CB33.bin.156             3.74  \n",
       "CBrest.bin.407           2.19  \n",
       "CB33.bin.101             2.06  \n",
       "CB33.bin.212             4.13  \n",
       "CB33.bin.45              2.13  \n",
       "CBrest.bin.395           3.26  \n",
       "CBrest.bin.40            1.89  \n",
       "CBrest.bin.368           2.65  \n",
       "CBrest.bin.358           3.56  \n",
       "CBrest.bin.246           1.96  \n",
       "CB33.bin.63              2.20  \n",
       "CBrest.bin.417           2.30  \n",
       "CB33.bin.12              1.96  \n",
       "CBrest.bin.188           2.84  \n",
       "CBrest.bin.408           3.33  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "first_bool = (qualtax['Completeness'] > 80.) & (qualtax[abund_srs.name] > 1)\n",
    "print(first_bool.sum(), 'completeish bins')\n",
    "\n",
    "display_cols = ['Completeness', 'Contamination', 'red_value', 'aa_percent', 'Phylum',\n",
    "                'Class', 'Order', 'Family', 'Genus', 'Species', abund_srs.name, 'MaxCopyNumber', \n",
    "                'N50 (contigs)', 'GenomeSizeMbp']\n",
    "\n",
    "\n",
    "qualtax.loc[first_bool, display_cols].sort_values(by='MaxCopyNumber', ascending=False).head(35)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "structured-turkish",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "167\n",
      "\n",
      "0.005988023952095809\n",
      "Unclassified Class\n",
      "\t Delongbacteria 1\n",
      "0.005988023952095809\n",
      "Unclassified Order\n",
      "\t UBA6777 1\n",
      "0.059880239520958084\n",
      "Unclassified Family\n",
      "\t Thiohalospirales 3\n",
      "\t Calditrichales 1\n",
      "\t Flavobacteriales 3\n",
      "\t Nanopelagicales 2\n",
      "\t Chthoniobacterales 1\n",
      "0.2634730538922156\n",
      "Unclassified Genus\n",
      "\t Pelagibacteraceae 2\n",
      "\t LZORAL124-64-63 2\n",
      "\t D2472 1\n",
      "\t SM1A02 3\n",
      "\t UBA955 2\n",
      "\t Ilumatobacteraceae 4\n",
      "\t Chromatiaceae 1\n",
      "\t UBA8477 2\n",
      "\t Thioglobaceae 2\n",
      "\t UBA3470 1\n",
      "\t envOPS12 1\n",
      "\t Phycisphaeraceae 2\n",
      "\t UBA8416 1\n",
      "\t Planctomycetaceae 2\n",
      "\t GCA-001735895 1\n",
      "\t S36-B12 1\n",
      "\t Sedimenticolaceae 1\n",
      "\t HTCC2089 3\n",
      "\t Magnetovibrionaceae 1\n",
      "\t Wenzhouxiangellaceae 1\n",
      "\t UBA8649 1\n",
      "\t UBA10348 1\n",
      "\t Flavobacteriaceae 2\n",
      "\t UBA2386 1\n",
      "\t Saprospiraceae 2\n",
      "\t UBA12015 1\n",
      "\t UBA1268 1\n",
      "\t Silvanigrellaceae 1\n",
      "0.5029940119760479\n",
      "Unclassified Species\n",
      "\t Nitrosopumilus 3\n",
      "\t UBA868 2\n",
      "\t UA16 2\n",
      "\t Casp-actino8 1\n",
      "\t UBA2463 4\n",
      "\t LZORAL124-64-63 3\n",
      "\t AAA164-E04 2\n",
      "\t Marivivens 3\n",
      "\t MAG-120802 4\n",
      "\t UBA4655 1\n",
      "\t Pontimonas 3\n",
      "\t IMCC3088 2\n",
      "\t Puniceispirillum 3\n",
      "\t BACL27 3\n",
      "\t GCF-002020805 1\n",
      "\t Casp-actino5 1\n",
      "\t MONJU 2\n",
      "\t UBA11426 1\n",
      "\t UBA964 2\n",
      "\t BACL21 2\n",
      "\t Pelagibacter 2\n",
      "\t RS62 2\n",
      "\t Sedimenticola 3\n",
      "\t UBA10416 1\n",
      "\t HIMB11 2\n",
      "\t Vulcanococcus 3\n",
      "\t BACL14 2\n",
      "\t IMCC9063 2\n",
      "\t Arctic95D-9 1\n",
      "\t D2472 1\n",
      "\t HC6-5 1\n",
      "\t UBA10649 1\n",
      "\t Haliscomenobacter 1\n",
      "\t Litoricola 1\n",
      "\t HIMB114 1\n",
      "\t AAA536-G10 1\n",
      "\t MED-G11 1\n",
      "\t Marivita 1\n",
      "\t ATZT02 2\n",
      "\t PCC7001 1\n",
      "\t UBA2329 1\n",
      "\t S36-B12 2\n",
      "\t Synechococcus 2\n",
      "\t Mariprofundus 1\n",
      "\t MS024-2A 1\n",
      "\t OM182 1\n",
      "\t Planktomarina 1\n"
     ]
    }
   ],
   "source": [
    "zeroth_bool = (qualtax['Completeness'] > 80.)\n",
    "print(zeroth_bool.sum())\n",
    "from collections import Counter\n",
    "tax_cols = ['Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species']\n",
    "\n",
    "print()\n",
    "\n",
    "for tc in tax_cols[1:]:\n",
    "    prev_col = tax_cols[(tax_cols.index(tc)-1)]\n",
    "    print(((qualtax.loc[zeroth_bool & (qualtax[tc] == ''), tc ] == '').sum() - (qualtax.loc[zeroth_bool & (qualtax[tc] == ''), prev_col ] == '').sum()) / zeroth_bool.sum())\n",
    "    unclassifieds = list(qualtax.loc[zeroth_bool & (qualtax[tc] == ''), prev_col])\n",
    "    \n",
    "    print('Unclassified', tc)\n",
    "    for i, j in Counter(unclassifieds).items():\n",
    "        if i != '':\n",
    "            print(\"\\t\", i, j)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "suitable-effect",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "184 genome annotations identified\n"
     ]
    }
   ],
   "source": [
    "sys.path.append('/Volumes/KeithSSD/SulFox/scripts')\n",
    "from pathwayFunctions import *\n",
    "\n",
    "base_path = \"/Volumes/KeithSSD/SulFox/data\"\n",
    "\n",
    "# Load abundances:\n",
    "bin_abundances = '/Volumes/KeithSSD/CBFunctions/data/outputs/bin_abund_n_metadata.p'\n",
    "cbTr_abnd2, cb33_abnd2, metadata = metadata_and_abundances(bin_abundances)\n",
    "\n",
    "# load lineage info & quality \n",
    "qualtax_file = '/Volumes/KeithSSD/CBFunctions/data/outputs/qual_tax_full.p'\n",
    "qualtax_df = load_qualtax(qualtax_file)\n",
    "\n",
    "# Locate prokka annotations \n",
    "annotation_folder = os.path.join(base_path, 'annotation')\n",
    "bins_preselected = locate_annotated_files(annotation_folder)\n",
    "bins_desired = [i for i in bins_preselected if i in set(qualtax_df.index[qualtax_df['Completeness'] >= 80])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "clear-western",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Completeness</th>\n",
       "      <th>Contamination</th>\n",
       "      <th>red_value</th>\n",
       "      <th>aa_percent</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>% samples w/ Mean Cov &gt; 2x</th>\n",
       "      <th>MaxCopyNumber</th>\n",
       "      <th>N50 (contigs)</th>\n",
       "      <th>GenomeSizeMbp</th>\n",
       "      <th>PFX</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CB33.bin.2</th>\n",
       "      <td>93.10</td>\n",
       "      <td>15.20</td>\n",
       "      <td>0.919</td>\n",
       "      <td>88.10</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Nanopelagicales</td>\n",
       "      <td>Nanopelagicaceae</td>\n",
       "      <td>MAG-120802</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>5.5</td>\n",
       "      <td>19441</td>\n",
       "      <td>1.77</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.188</th>\n",
       "      <td>98.12</td>\n",
       "      <td>19.89</td>\n",
       "      <td>NaN</td>\n",
       "      <td>98.37</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Flavobacteriales</td>\n",
       "      <td>Cryomorphaceae</td>\n",
       "      <td>TMED14</td>\n",
       "      <td>TMED14 sp002381225</td>\n",
       "      <td>2</td>\n",
       "      <td>1.8</td>\n",
       "      <td>36065</td>\n",
       "      <td>2.84</td>\n",
       "      <td>CBrest</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.40</th>\n",
       "      <td>83.53</td>\n",
       "      <td>3.67</td>\n",
       "      <td>0.969</td>\n",
       "      <td>79.80</td>\n",
       "      <td>Bacteroidota</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Flavobacteriales</td>\n",
       "      <td>Flavobacteriaceae</td>\n",
       "      <td>MS024-2A</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>2.2</td>\n",
       "      <td>86578</td>\n",
       "      <td>1.89</td>\n",
       "      <td>CBrest</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.84</th>\n",
       "      <td>94.75</td>\n",
       "      <td>21.07</td>\n",
       "      <td>0.990</td>\n",
       "      <td>79.63</td>\n",
       "      <td>Crenarchaeota</td>\n",
       "      <td>Nitrososphaeria</td>\n",
       "      <td>Nitrososphaerales</td>\n",
       "      <td>Nitrosopumilaceae</td>\n",
       "      <td>Nitrosopumilus</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>1.6</td>\n",
       "      <td>9146</td>\n",
       "      <td>2.21</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.104</th>\n",
       "      <td>97.09</td>\n",
       "      <td>29.45</td>\n",
       "      <td>0.991</td>\n",
       "      <td>85.48</td>\n",
       "      <td>Crenarchaeota</td>\n",
       "      <td>Nitrososphaeria</td>\n",
       "      <td>Nitrososphaerales</td>\n",
       "      <td>Nitrosopumilaceae</td>\n",
       "      <td>Nitrosopumilus</td>\n",
       "      <td></td>\n",
       "      <td>6</td>\n",
       "      <td>7.1</td>\n",
       "      <td>21735</td>\n",
       "      <td>1.76</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.246</th>\n",
       "      <td>89.20</td>\n",
       "      <td>13.59</td>\n",
       "      <td>0.989</td>\n",
       "      <td>76.83</td>\n",
       "      <td>Crenarchaeota</td>\n",
       "      <td>Nitrososphaeria</td>\n",
       "      <td>Nitrososphaerales</td>\n",
       "      <td>Nitrosopumilaceae</td>\n",
       "      <td>Nitrosopumilus</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>13630</td>\n",
       "      <td>1.96</td>\n",
       "      <td>CBrest</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.408</th>\n",
       "      <td>99.34</td>\n",
       "      <td>51.18</td>\n",
       "      <td>0.905</td>\n",
       "      <td>93.73</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteriia</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>Cyanobiaceae</td>\n",
       "      <td>Synechococcus</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>1.7</td>\n",
       "      <td>30189</td>\n",
       "      <td>3.33</td>\n",
       "      <td>CBrest</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.63</th>\n",
       "      <td>91.99</td>\n",
       "      <td>13.27</td>\n",
       "      <td>0.951</td>\n",
       "      <td>82.50</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteriia</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>Cyanobiaceae</td>\n",
       "      <td>Vulcanococcus</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6759</td>\n",
       "      <td>2.20</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.111</th>\n",
       "      <td>84.74</td>\n",
       "      <td>18.58</td>\n",
       "      <td>NaN</td>\n",
       "      <td>72.22</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Cyanobacteriia</td>\n",
       "      <td>Synechococcales</td>\n",
       "      <td>Cyanobiaceae</td>\n",
       "      <td>Vulcanococcus</td>\n",
       "      <td>Vulcanococcus sp000179255</td>\n",
       "      <td>3</td>\n",
       "      <td>2.6</td>\n",
       "      <td>5654</td>\n",
       "      <td>2.24</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.207</th>\n",
       "      <td>94.51</td>\n",
       "      <td>9.07</td>\n",
       "      <td>0.396</td>\n",
       "      <td>91.37</td>\n",
       "      <td>Delongbacteria</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>1.7</td>\n",
       "      <td>16374</td>\n",
       "      <td>3.31</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.179</th>\n",
       "      <td>88.81</td>\n",
       "      <td>8.11</td>\n",
       "      <td>0.889</td>\n",
       "      <td>79.58</td>\n",
       "      <td>Krumholzibacteriota</td>\n",
       "      <td>Krumholzibacteria</td>\n",
       "      <td>LZORAL124-64-63</td>\n",
       "      <td>LZORAL124-64-63</td>\n",
       "      <td>LZORAL124-64-63</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>1.2</td>\n",
       "      <td>8614</td>\n",
       "      <td>4.59</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.229</th>\n",
       "      <td>96.70</td>\n",
       "      <td>21.44</td>\n",
       "      <td>0.708</td>\n",
       "      <td>94.64</td>\n",
       "      <td>Marinisomatota</td>\n",
       "      <td>UBA8477</td>\n",
       "      <td>UBA8477</td>\n",
       "      <td>UBA8477</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>1.6</td>\n",
       "      <td>36449</td>\n",
       "      <td>4.60</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.182</th>\n",
       "      <td>90.11</td>\n",
       "      <td>9.89</td>\n",
       "      <td>0.712</td>\n",
       "      <td>91.73</td>\n",
       "      <td>Marinisomatota</td>\n",
       "      <td>UBA8477</td>\n",
       "      <td>UBA8477</td>\n",
       "      <td>UBA8477</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5</td>\n",
       "      <td>6.1</td>\n",
       "      <td>85486</td>\n",
       "      <td>4.43</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.167</th>\n",
       "      <td>84.99</td>\n",
       "      <td>12.74</td>\n",
       "      <td>0.523</td>\n",
       "      <td>88.87</td>\n",
       "      <td>Myxococcota</td>\n",
       "      <td>UBA6777</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>2.6</td>\n",
       "      <td>10422</td>\n",
       "      <td>6.35</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.95</th>\n",
       "      <td>96.55</td>\n",
       "      <td>15.05</td>\n",
       "      <td>0.960</td>\n",
       "      <td>95.38</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rhodobacterales</td>\n",
       "      <td>Rhodobacteraceae</td>\n",
       "      <td>HIMB11</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>1.4</td>\n",
       "      <td>18223</td>\n",
       "      <td>3.37</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.222</th>\n",
       "      <td>96.52</td>\n",
       "      <td>5.91</td>\n",
       "      <td>NaN</td>\n",
       "      <td>95.10</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rhodobacterales</td>\n",
       "      <td>Rhodobacteraceae</td>\n",
       "      <td>HIMB11</td>\n",
       "      <td>HIMB11 sp001642945</td>\n",
       "      <td>2</td>\n",
       "      <td>1.2</td>\n",
       "      <td>10085</td>\n",
       "      <td>3.20</td>\n",
       "      <td>CBrest</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.174</th>\n",
       "      <td>85.92</td>\n",
       "      <td>27.07</td>\n",
       "      <td>0.940</td>\n",
       "      <td>85.22</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rhodobacterales</td>\n",
       "      <td>Rhodobacteraceae</td>\n",
       "      <td>Marivivens</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>2.5</td>\n",
       "      <td>13294</td>\n",
       "      <td>3.41</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.358</th>\n",
       "      <td>91.38</td>\n",
       "      <td>9.72</td>\n",
       "      <td>0.940</td>\n",
       "      <td>94.84</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rhodobacterales</td>\n",
       "      <td>Rhodobacteraceae</td>\n",
       "      <td>Marivivens</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>2.1</td>\n",
       "      <td>57260</td>\n",
       "      <td>3.56</td>\n",
       "      <td>CBrest</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.287</th>\n",
       "      <td>97.20</td>\n",
       "      <td>6.47</td>\n",
       "      <td>NaN</td>\n",
       "      <td>93.85</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rhodobacterales</td>\n",
       "      <td>Rhodobacteraceae</td>\n",
       "      <td>Paracoccus</td>\n",
       "      <td>Paracoccus sp000967825</td>\n",
       "      <td>2</td>\n",
       "      <td>1.4</td>\n",
       "      <td>33054</td>\n",
       "      <td>4.46</td>\n",
       "      <td>CBrest</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.233</th>\n",
       "      <td>93.79</td>\n",
       "      <td>34.78</td>\n",
       "      <td>0.885</td>\n",
       "      <td>91.51</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Chromatiales</td>\n",
       "      <td>Sedimenticolaceae</td>\n",
       "      <td>MONJU</td>\n",
       "      <td></td>\n",
       "      <td>4</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4151</td>\n",
       "      <td>2.28</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.368</th>\n",
       "      <td>97.23</td>\n",
       "      <td>26.67</td>\n",
       "      <td>0.884</td>\n",
       "      <td>97.30</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Chromatiales</td>\n",
       "      <td>Sedimenticolaceae</td>\n",
       "      <td>MONJU</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>2.1</td>\n",
       "      <td>14377</td>\n",
       "      <td>2.65</td>\n",
       "      <td>CBrest</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.236</th>\n",
       "      <td>95.35</td>\n",
       "      <td>7.26</td>\n",
       "      <td>0.931</td>\n",
       "      <td>89.09</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Chromatiales</td>\n",
       "      <td>Sedimenticolaceae</td>\n",
       "      <td>Sedimenticola</td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>4.3</td>\n",
       "      <td>21011</td>\n",
       "      <td>3.80</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.395</th>\n",
       "      <td>85.05</td>\n",
       "      <td>1.80</td>\n",
       "      <td>0.931</td>\n",
       "      <td>82.94</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Chromatiales</td>\n",
       "      <td>Sedimenticolaceae</td>\n",
       "      <td>Sedimenticola</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>2.2</td>\n",
       "      <td>52182</td>\n",
       "      <td>3.26</td>\n",
       "      <td>CBrest</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.196</th>\n",
       "      <td>92.48</td>\n",
       "      <td>12.41</td>\n",
       "      <td>0.820</td>\n",
       "      <td>90.08</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>GCA-001735895</td>\n",
       "      <td>GCA-001735895</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>2.7</td>\n",
       "      <td>24647</td>\n",
       "      <td>5.29</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.169</th>\n",
       "      <td>91.35</td>\n",
       "      <td>20.63</td>\n",
       "      <td>0.992</td>\n",
       "      <td>90.63</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Halieaceae</td>\n",
       "      <td>IMCC3088</td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>1.3</td>\n",
       "      <td>5824</td>\n",
       "      <td>3.03</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.149</th>\n",
       "      <td>97.90</td>\n",
       "      <td>95.24</td>\n",
       "      <td>0.688</td>\n",
       "      <td>93.08</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiohalospirales</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>6</td>\n",
       "      <td>4.4</td>\n",
       "      <td>15544</td>\n",
       "      <td>3.09</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.375</th>\n",
       "      <td>88.39</td>\n",
       "      <td>9.98</td>\n",
       "      <td>0.687</td>\n",
       "      <td>92.94</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiohalospirales</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>14.5</td>\n",
       "      <td>17860</td>\n",
       "      <td>1.56</td>\n",
       "      <td>CBrest</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.191</th>\n",
       "      <td>88.97</td>\n",
       "      <td>2.97</td>\n",
       "      <td>0.819</td>\n",
       "      <td>89.52</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiomicrospirales</td>\n",
       "      <td>Thioglobaceae</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>2.8</td>\n",
       "      <td>8445</td>\n",
       "      <td>1.26</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.263</th>\n",
       "      <td>90.14</td>\n",
       "      <td>17.11</td>\n",
       "      <td>0.818</td>\n",
       "      <td>88.27</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Thiomicrospirales</td>\n",
       "      <td>Thioglobaceae</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2</td>\n",
       "      <td>5.4</td>\n",
       "      <td>10418</td>\n",
       "      <td>1.87</td>\n",
       "      <td>CBrest</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CBrest.bin.394</th>\n",
       "      <td>94.51</td>\n",
       "      <td>16.77</td>\n",
       "      <td>0.931</td>\n",
       "      <td>95.77</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>UBA10353</td>\n",
       "      <td>LS-SOB</td>\n",
       "      <td>UBA868</td>\n",
       "      <td></td>\n",
       "      <td>5</td>\n",
       "      <td>1.6</td>\n",
       "      <td>48629</td>\n",
       "      <td>2.79</td>\n",
       "      <td>CBrest</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.133</th>\n",
       "      <td>85.38</td>\n",
       "      <td>4.12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>83.25</td>\n",
       "      <td>SAR324</td>\n",
       "      <td>SAR324</td>\n",
       "      <td>SAR324</td>\n",
       "      <td>NAC60-12</td>\n",
       "      <td>JCVI-SCAAA005</td>\n",
       "      <td>JCVI-SCAAA005 sp002450295</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6090</td>\n",
       "      <td>4.85</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CB33.bin.212</th>\n",
       "      <td>92.40</td>\n",
       "      <td>5.73</td>\n",
       "      <td>0.704</td>\n",
       "      <td>86.75</td>\n",
       "      <td>Verrucomicrobiota</td>\n",
       "      <td>Kiritimatiellae</td>\n",
       "      <td>UBA8416</td>\n",
       "      <td>UBA8416</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3</td>\n",
       "      <td>2.3</td>\n",
       "      <td>12322</td>\n",
       "      <td>4.13</td>\n",
       "      <td>CB33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                Completeness  Contamination  red_value  aa_percent  \\\n",
       "CB33.bin.2             93.10          15.20      0.919       88.10   \n",
       "CBrest.bin.188         98.12          19.89        NaN       98.37   \n",
       "CBrest.bin.40          83.53           3.67      0.969       79.80   \n",
       "CB33.bin.84            94.75          21.07      0.990       79.63   \n",
       "CB33.bin.104           97.09          29.45      0.991       85.48   \n",
       "CBrest.bin.246         89.20          13.59      0.989       76.83   \n",
       "CBrest.bin.408         99.34          51.18      0.905       93.73   \n",
       "CB33.bin.63            91.99          13.27      0.951       82.50   \n",
       "CB33.bin.111           84.74          18.58        NaN       72.22   \n",
       "CB33.bin.207           94.51           9.07      0.396       91.37   \n",
       "CB33.bin.179           88.81           8.11      0.889       79.58   \n",
       "CB33.bin.229           96.70          21.44      0.708       94.64   \n",
       "CB33.bin.182           90.11           9.89      0.712       91.73   \n",
       "CB33.bin.167           84.99          12.74      0.523       88.87   \n",
       "CB33.bin.95            96.55          15.05      0.960       95.38   \n",
       "CBrest.bin.222         96.52           5.91        NaN       95.10   \n",
       "CB33.bin.174           85.92          27.07      0.940       85.22   \n",
       "CBrest.bin.358         91.38           9.72      0.940       94.84   \n",
       "CBrest.bin.287         97.20           6.47        NaN       93.85   \n",
       "CB33.bin.233           93.79          34.78      0.885       91.51   \n",
       "CBrest.bin.368         97.23          26.67      0.884       97.30   \n",
       "CB33.bin.236           95.35           7.26      0.931       89.09   \n",
       "CBrest.bin.395         85.05           1.80      0.931       82.94   \n",
       "CB33.bin.196           92.48          12.41      0.820       90.08   \n",
       "CB33.bin.169           91.35          20.63      0.992       90.63   \n",
       "CB33.bin.149           97.90          95.24      0.688       93.08   \n",
       "CBrest.bin.375         88.39           9.98      0.687       92.94   \n",
       "CB33.bin.191           88.97           2.97      0.819       89.52   \n",
       "CBrest.bin.263         90.14          17.11      0.818       88.27   \n",
       "CBrest.bin.394         94.51          16.77      0.931       95.77   \n",
       "CB33.bin.133           85.38           4.12        NaN       83.25   \n",
       "CB33.bin.212           92.40           5.73      0.704       86.75   \n",
       "\n",
       "                             Phylum                Class              Order  \\\n",
       "CB33.bin.2         Actinobacteriota       Actinobacteria    Nanopelagicales   \n",
       "CBrest.bin.188         Bacteroidota          Bacteroidia   Flavobacteriales   \n",
       "CBrest.bin.40          Bacteroidota          Bacteroidia   Flavobacteriales   \n",
       "CB33.bin.84           Crenarchaeota      Nitrososphaeria  Nitrososphaerales   \n",
       "CB33.bin.104          Crenarchaeota      Nitrososphaeria  Nitrososphaerales   \n",
       "CBrest.bin.246        Crenarchaeota      Nitrososphaeria  Nitrososphaerales   \n",
       "CBrest.bin.408        Cyanobacteria       Cyanobacteriia    Synechococcales   \n",
       "CB33.bin.63           Cyanobacteria       Cyanobacteriia    Synechococcales   \n",
       "CB33.bin.111          Cyanobacteria       Cyanobacteriia    Synechococcales   \n",
       "CB33.bin.207         Delongbacteria                                           \n",
       "CB33.bin.179    Krumholzibacteriota    Krumholzibacteria    LZORAL124-64-63   \n",
       "CB33.bin.229         Marinisomatota              UBA8477            UBA8477   \n",
       "CB33.bin.182         Marinisomatota              UBA8477            UBA8477   \n",
       "CB33.bin.167            Myxococcota              UBA6777                      \n",
       "CB33.bin.95          Proteobacteria  Alphaproteobacteria    Rhodobacterales   \n",
       "CBrest.bin.222       Proteobacteria  Alphaproteobacteria    Rhodobacterales   \n",
       "CB33.bin.174         Proteobacteria  Alphaproteobacteria    Rhodobacterales   \n",
       "CBrest.bin.358       Proteobacteria  Alphaproteobacteria    Rhodobacterales   \n",
       "CBrest.bin.287       Proteobacteria  Alphaproteobacteria    Rhodobacterales   \n",
       "CB33.bin.233         Proteobacteria  Gammaproteobacteria       Chromatiales   \n",
       "CBrest.bin.368       Proteobacteria  Gammaproteobacteria       Chromatiales   \n",
       "CB33.bin.236         Proteobacteria  Gammaproteobacteria       Chromatiales   \n",
       "CBrest.bin.395       Proteobacteria  Gammaproteobacteria       Chromatiales   \n",
       "CB33.bin.196         Proteobacteria  Gammaproteobacteria      GCA-001735895   \n",
       "CB33.bin.169         Proteobacteria  Gammaproteobacteria    Pseudomonadales   \n",
       "CB33.bin.149         Proteobacteria  Gammaproteobacteria   Thiohalospirales   \n",
       "CBrest.bin.375       Proteobacteria  Gammaproteobacteria   Thiohalospirales   \n",
       "CB33.bin.191         Proteobacteria  Gammaproteobacteria  Thiomicrospirales   \n",
       "CBrest.bin.263       Proteobacteria  Gammaproteobacteria  Thiomicrospirales   \n",
       "CBrest.bin.394       Proteobacteria  Gammaproteobacteria           UBA10353   \n",
       "CB33.bin.133                 SAR324               SAR324             SAR324   \n",
       "CB33.bin.212      Verrucomicrobiota      Kiritimatiellae            UBA8416   \n",
       "\n",
       "                           Family            Genus                    Species  \\\n",
       "CB33.bin.2       Nanopelagicaceae       MAG-120802                              \n",
       "CBrest.bin.188     Cryomorphaceae           TMED14         TMED14 sp002381225   \n",
       "CBrest.bin.40   Flavobacteriaceae         MS024-2A                              \n",
       "CB33.bin.84     Nitrosopumilaceae   Nitrosopumilus                              \n",
       "CB33.bin.104    Nitrosopumilaceae   Nitrosopumilus                              \n",
       "CBrest.bin.246  Nitrosopumilaceae   Nitrosopumilus                              \n",
       "CBrest.bin.408       Cyanobiaceae    Synechococcus                              \n",
       "CB33.bin.63          Cyanobiaceae    Vulcanococcus                              \n",
       "CB33.bin.111         Cyanobiaceae    Vulcanococcus  Vulcanococcus sp000179255   \n",
       "CB33.bin.207                                                                    \n",
       "CB33.bin.179      LZORAL124-64-63  LZORAL124-64-63                              \n",
       "CB33.bin.229              UBA8477                                               \n",
       "CB33.bin.182              UBA8477                                               \n",
       "CB33.bin.167                                                                    \n",
       "CB33.bin.95      Rhodobacteraceae           HIMB11                              \n",
       "CBrest.bin.222   Rhodobacteraceae           HIMB11         HIMB11 sp001642945   \n",
       "CB33.bin.174     Rhodobacteraceae       Marivivens                              \n",
       "CBrest.bin.358   Rhodobacteraceae       Marivivens                              \n",
       "CBrest.bin.287   Rhodobacteraceae       Paracoccus     Paracoccus sp000967825   \n",
       "CB33.bin.233    Sedimenticolaceae            MONJU                              \n",
       "CBrest.bin.368  Sedimenticolaceae            MONJU                              \n",
       "CB33.bin.236    Sedimenticolaceae    Sedimenticola                              \n",
       "CBrest.bin.395  Sedimenticolaceae    Sedimenticola                              \n",
       "CB33.bin.196        GCA-001735895                                               \n",
       "CB33.bin.169           Halieaceae         IMCC3088                              \n",
       "CB33.bin.149                                                                    \n",
       "CBrest.bin.375                                                                  \n",
       "CB33.bin.191        Thioglobaceae                                               \n",
       "CBrest.bin.263      Thioglobaceae                                               \n",
       "CBrest.bin.394             LS-SOB           UBA868                              \n",
       "CB33.bin.133             NAC60-12    JCVI-SCAAA005  JCVI-SCAAA005 sp002450295   \n",
       "CB33.bin.212              UBA8416                                               \n",
       "\n",
       "                % samples w/ Mean Cov > 2x  MaxCopyNumber N50 (contigs)  \\\n",
       "CB33.bin.2                               4            5.5         19441   \n",
       "CBrest.bin.188                           2            1.8         36065   \n",
       "CBrest.bin.40                            3            2.2         86578   \n",
       "CB33.bin.84                              3            1.6          9146   \n",
       "CB33.bin.104                             6            7.1         21735   \n",
       "CBrest.bin.246                           2            2.0         13630   \n",
       "CBrest.bin.408                           3            1.7         30189   \n",
       "CB33.bin.63                              2            2.0          6759   \n",
       "CB33.bin.111                             3            2.6          5654   \n",
       "CB33.bin.207                             4            1.7         16374   \n",
       "CB33.bin.179                             2            1.2          8614   \n",
       "CB33.bin.229                             4            1.6         36449   \n",
       "CB33.bin.182                             5            6.1         85486   \n",
       "CB33.bin.167                             3            2.6         10422   \n",
       "CB33.bin.95                              2            1.4         18223   \n",
       "CBrest.bin.222                           2            1.2         10085   \n",
       "CB33.bin.174                             2            2.5         13294   \n",
       "CBrest.bin.358                           3            2.1         57260   \n",
       "CBrest.bin.287                           2            1.4         33054   \n",
       "CB33.bin.233                             4            5.0          4151   \n",
       "CBrest.bin.368                           3            2.1         14377   \n",
       "CB33.bin.236                             3            4.3         21011   \n",
       "CBrest.bin.395                           2            2.2         52182   \n",
       "CB33.bin.196                             2            2.7         24647   \n",
       "CB33.bin.169                             2            1.3          5824   \n",
       "CB33.bin.149                             6            4.4         15544   \n",
       "CBrest.bin.375                           3           14.5         17860   \n",
       "CB33.bin.191                             2            2.8          8445   \n",
       "CBrest.bin.263                           2            5.4         10418   \n",
       "CBrest.bin.394                           5            1.6         48629   \n",
       "CB33.bin.133                             2            1.0          6090   \n",
       "CB33.bin.212                             3            2.3         12322   \n",
       "\n",
       "                GenomeSizeMbp     PFX  \n",
       "CB33.bin.2               1.77    CB33  \n",
       "CBrest.bin.188           2.84  CBrest  \n",
       "CBrest.bin.40            1.89  CBrest  \n",
       "CB33.bin.84              2.21    CB33  \n",
       "CB33.bin.104             1.76    CB33  \n",
       "CBrest.bin.246           1.96  CBrest  \n",
       "CBrest.bin.408           3.33  CBrest  \n",
       "CB33.bin.63              2.20    CB33  \n",
       "CB33.bin.111             2.24    CB33  \n",
       "CB33.bin.207             3.31    CB33  \n",
       "CB33.bin.179             4.59    CB33  \n",
       "CB33.bin.229             4.60    CB33  \n",
       "CB33.bin.182             4.43    CB33  \n",
       "CB33.bin.167             6.35    CB33  \n",
       "CB33.bin.95              3.37    CB33  \n",
       "CBrest.bin.222           3.20  CBrest  \n",
       "CB33.bin.174             3.41    CB33  \n",
       "CBrest.bin.358           3.56  CBrest  \n",
       "CBrest.bin.287           4.46  CBrest  \n",
       "CB33.bin.233             2.28    CB33  \n",
       "CBrest.bin.368           2.65  CBrest  \n",
       "CB33.bin.236             3.80    CB33  \n",
       "CBrest.bin.395           3.26  CBrest  \n",
       "CB33.bin.196             5.29    CB33  \n",
       "CB33.bin.169             3.03    CB33  \n",
       "CB33.bin.149             3.09    CB33  \n",
       "CBrest.bin.375           1.56  CBrest  \n",
       "CB33.bin.191             1.26    CB33  \n",
       "CBrest.bin.263           1.87  CBrest  \n",
       "CBrest.bin.394           2.79  CBrest  \n",
       "CB33.bin.133             4.85    CB33  \n",
       "CB33.bin.212             4.13    CB33  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "qualtax['PFX'] = pd.Series({i:i.split(\".\")[0] for i in qualtax.index})\n",
    "sort_columns = ['Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species', 'PFX', 'MaxCopyNumber']\n",
    "bins_of_interest = ['CBrest.bin.263', 'CB33.bin.191', 'CBrest.bin.246', 'CBrest.bin.222', 'CBrest.bin.368', \n",
    "                    'CBrest.bin.375', 'CB33.bin.236', 'CBrest.bin.358', 'CB33.bin.182', 'CB33.bin.2', 'CBrest.bin.408']\n",
    "qualtax.loc[(set(bins_desired) & set(qualtax.index[first_bool])), display_cols + ['PFX']].sort_values(by=sort_columns)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "coastal-librarian",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\tCB33.bin.104 pathways discovered: 118\n",
      "26 pathway associated proteins (of interest)\n",
      "\tCB33.bin.111 pathways discovered: 175\n",
      "53 pathway associated proteins (of interest)\n",
      "\tCB33.bin.118 pathways discovered: 240\n",
      "46 pathway associated proteins (of interest)\n",
      "\tCB33.bin.133 pathways discovered: 233\n",
      "45 pathway associated proteins (of interest)\n",
      "\tCB33.bin.142 pathways discovered: 213\n",
      "35 pathway associated proteins (of interest)\n",
      "\tCB33.bin.146 pathways discovered: 180\n",
      "48 pathway associated proteins (of interest)\n",
      "\tCB33.bin.149 pathways discovered: 153\n",
      "79 pathway associated proteins (of interest)\n",
      "\tCB33.bin.167 pathways discovered: 204\n",
      "46 pathway associated proteins (of interest)\n",
      "\tCB33.bin.168 pathways discovered: 200\n",
      "30 pathway associated proteins (of interest)\n",
      "\tCB33.bin.169 pathways discovered: 204\n",
      "35 pathway associated proteins (of interest)\n",
      "\tCB33.bin.172 pathways discovered: 162\n",
      "19 pathway associated proteins (of interest)\n",
      "\tCB33.bin.174 pathways discovered: 236\n",
      "72 pathway associated proteins (of interest)\n",
      "\tCB33.bin.175 pathways discovered: 240\n",
      "51 pathway associated proteins (of interest)\n",
      "\tCB33.bin.179 pathways discovered: 188\n",
      "40 pathway associated proteins (of interest)\n",
      "\tCB33.bin.182 pathways discovered: 184\n",
      "31 pathway associated proteins (of interest)\n",
      "\tCB33.bin.185 pathways discovered: 210\n",
      "43 pathway associated proteins (of interest)\n",
      "\tCB33.bin.186 pathways discovered: 228\n",
      "79 pathway associated proteins (of interest)\n",
      "\tCB33.bin.190 pathways discovered: 185\n",
      "39 pathway associated proteins (of interest)\n",
      "\tCB33.bin.191 pathways discovered: 119\n",
      "44 pathway associated proteins (of interest)\n",
      "\tCB33.bin.194 pathways discovered: 145\n",
      "43 pathway associated proteins (of interest)\n",
      "\tCB33.bin.196 pathways discovered: 204\n",
      "77 pathway associated proteins (of interest)\n",
      "\tCB33.bin.2 pathways discovered: 138\n",
      "22 pathway associated proteins (of interest)\n",
      "\tCB33.bin.201 pathways discovered: 200\n",
      "24 pathway associated proteins (of interest)\n",
      "\tCB33.bin.202 pathways discovered: 195\n",
      "70 pathway associated proteins (of interest)\n",
      "\tCB33.bin.203 pathways discovered: 205\n",
      "32 pathway associated proteins (of interest)\n",
      "\tCB33.bin.207 pathways discovered: 161\n",
      "20 pathway associated proteins (of interest)\n",
      "\tCB33.bin.212 pathways discovered: 179\n",
      "20 pathway associated proteins (of interest)\n",
      "\tCB33.bin.216 pathways discovered: 169\n",
      "71 pathway associated proteins (of interest)\n",
      "\tCB33.bin.218 pathways discovered: 181\n",
      "42 pathway associated proteins (of interest)\n",
      "\tCB33.bin.219 pathways discovered: 252\n",
      "34 pathway associated proteins (of interest)\n",
      "\tCB33.bin.222 pathways discovered: 233\n",
      "62 pathway associated proteins (of interest)\n",
      "\tCB33.bin.227 pathways discovered: 179\n",
      "25 pathway associated proteins (of interest)\n",
      "\tCB33.bin.229 pathways discovered: 186\n",
      "33 pathway associated proteins (of interest)\n",
      "\tCB33.bin.233 pathways discovered: 151\n",
      "47 pathway associated proteins (of interest)\n",
      "\tCB33.bin.235 pathways discovered: 209\n",
      "68 pathway associated proteins (of interest)\n",
      "\tCB33.bin.236 pathways discovered: 200\n",
      "85 pathway associated proteins (of interest)\n",
      "\tCB33.bin.32 pathways discovered: 164\n",
      "51 pathway associated proteins (of interest)\n",
      "\tCB33.bin.34 pathways discovered: 144\n",
      "53 pathway associated proteins (of interest)\n",
      "\tCB33.bin.56 pathways discovered: 214\n",
      "70 pathway associated proteins (of interest)\n",
      "\tCB33.bin.61 pathways discovered: 161\n",
      "34 pathway associated proteins (of interest)\n",
      "\tCB33.bin.63 pathways discovered: 183\n",
      "51 pathway associated proteins (of interest)\n",
      "\tCB33.bin.82 pathways discovered: 182\n",
      "24 pathway associated proteins (of interest)\n",
      "\tCB33.bin.84 pathways discovered: 122\n",
      "24 pathway associated proteins (of interest)\n",
      "\tCB33.bin.85 pathways discovered: 160\n",
      "19 pathway associated proteins (of interest)\n",
      "\tCB33.bin.86 pathways discovered: 162\n",
      "21 pathway associated proteins (of interest)\n",
      "\tCB33.bin.89 pathways discovered: 222\n",
      "45 pathway associated proteins (of interest)\n",
      "\tCB33.bin.90 pathways discovered: 226\n",
      "42 pathway associated proteins (of interest)\n",
      "\tCB33.bin.95 pathways discovered: 247\n",
      "64 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.111 pathways discovered: 154\n",
      "18 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.113 pathways discovered: 214\n",
      "71 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.118 pathways discovered: 188\n",
      "35 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.142 pathways discovered: 173\n",
      "20 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.174 pathways discovered: 239\n",
      "80 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.188 pathways discovered: 153\n",
      "18 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.189 pathways discovered: 182\n",
      "54 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.191 pathways discovered: 203\n",
      "37 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.196 pathways discovered: 221\n",
      "84 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.222 pathways discovered: 227\n",
      "40 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.223 pathways discovered: 180\n",
      "35 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.246 pathways discovered: 121\n",
      "23 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.25 pathways discovered: 150\n",
      "18 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.252 pathways discovered: 187\n",
      "56 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.26 pathways discovered: 296\n",
      "122 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.263 pathways discovered: 117\n",
      "50 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.287 pathways discovered: 259\n",
      "61 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.322 pathways discovered: 206\n",
      "35 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.327 pathways discovered: 202\n",
      "40 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.331 pathways discovered: 219\n",
      "52 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.333 pathways discovered: 231\n",
      "46 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.344 pathways discovered: 248\n",
      "66 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.346 pathways discovered: 181\n",
      "28 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.349 pathways discovered: 197\n",
      "69 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.355 pathways discovered: 154\n",
      "20 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.358 pathways discovered: 228\n",
      "62 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.359 pathways discovered: 195\n",
      "67 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.368 pathways discovered: 159\n",
      "51 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.370 pathways discovered: 210\n",
      "33 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.374 pathways discovered: 260\n",
      "58 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.375 pathways discovered: 139\n",
      "37 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.377 pathways discovered: 226\n",
      "44 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.381 pathways discovered: 150\n",
      "43 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.384 pathways discovered: 193\n",
      "24 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.385 pathways discovered: 224\n",
      "35 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.388 pathways discovered: 209\n",
      "75 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.389 pathways discovered: 184\n",
      "96 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.390 pathways discovered: 176\n",
      "32 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.394 pathways discovered: 234\n",
      "48 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.395 pathways discovered: 189\n",
      "73 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.40 pathways discovered: 163\n",
      "21 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.405 pathways discovered: 219\n",
      "42 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.408 pathways discovered: 203\n",
      "77 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.6 pathways discovered: 225\n",
      "42 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.61 pathways discovered: 248\n",
      "45 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.63 pathways discovered: 200\n",
      "28 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.84 pathways discovered: 207\n",
      "43 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.94 pathways discovered: 145\n",
      "18 pathway associated proteins (of interest)\n",
      "\tCBrest.bin.96 pathways discovered: 201\n",
      "30 pathway associated proteins (of interest)\n"
     ]
    }
   ],
   "source": [
    "# \n",
    "report_files_path = os.path.join(base_path, \"minpath_out\")\n",
    "target_pathways = os.path.join(base_path, \"tab_seperated_files/PathwaysConsidered.csv\")\n",
    "suffix = '.report.txt'\n",
    "mc_pathway_df = aggregate_and_filter_minpath_data(report_files_path, suffix, annotation_folder, target_pathways, bins_desired)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "danish-magic",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "97\n",
      "97\n"
     ]
    }
   ],
   "source": [
    "qt_idx = set(qualtax_df.index).intersection(set(mc_pathway_df.genome))\n",
    "complete_list = qt_idx - set(qualtax_df.index[qualtax_df['Completeness'] < 80])\n",
    "print(len(complete_list))\n",
    "print(len(qt_idx))\n",
    "mc_pathway_df = mc_pathway_df[mc_pathway_df.locus_tag.notnull() & mc_pathway_df.genome.isin(complete_list)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cathedral-collectible",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>product</th>\n",
       "      <th>eC_number</th>\n",
       "      <th>gene</th>\n",
       "      <th>db_xref</th>\n",
       "      <th>path_accession</th>\n",
       "      <th>path_name</th>\n",
       "      <th>metacyc_target</th>\n",
       "      <th>genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CB33_1618</td>\n",
       "      <td>1051</td>\n",
       "      <td>1809</td>\n",
       "      <td>-</td>\n",
       "      <td>HIBBKILN_00002</td>\n",
       "      <td>rhomboid family intramembrane serine protease</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>CB33.bin.104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CB33_1715</td>\n",
       "      <td>109</td>\n",
       "      <td>261</td>\n",
       "      <td>+</td>\n",
       "      <td>HIBBKILN_00005</td>\n",
       "      <td>S-adenosylmethionine synthetase</td>\n",
       "      <td>2.5.1.6</td>\n",
       "      <td>metK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PWY-6292</td>\n",
       "      <td>superpathway of L-cysteine biosynthesis (mamma...</td>\n",
       "      <td>False</td>\n",
       "      <td>CB33.bin.104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CB33_1715</td>\n",
       "      <td>1197</td>\n",
       "      <td>2282</td>\n",
       "      <td>+</td>\n",
       "      <td>HIBBKILN_00007</td>\n",
       "      <td>putative cobalt-precorrin-6A synthase (deacety...</td>\n",
       "      <td>2.1.1.-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PWY-6146</td>\n",
       "      <td>&lt;i&gt;Methanobacterium thermoautotrophicum&lt;/i&gt; bi...</td>\n",
       "      <td>False</td>\n",
       "      <td>CB33.bin.104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CB33_1715</td>\n",
       "      <td>2314</td>\n",
       "      <td>3366</td>\n",
       "      <td>+</td>\n",
       "      <td>HIBBKILN_00008</td>\n",
       "      <td>Uroporphyrin-III C-methyltransferase</td>\n",
       "      <td>2.1.1.107</td>\n",
       "      <td>cbiG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PWY-5196</td>\n",
       "      <td>factor 430 biosynthesis</td>\n",
       "      <td>False</td>\n",
       "      <td>CB33.bin.104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CB33_1715</td>\n",
       "      <td>4105</td>\n",
       "      <td>4746</td>\n",
       "      <td>+</td>\n",
       "      <td>HIBBKILN_00010</td>\n",
       "      <td>Cobalt-precorrin-8X methylmutase</td>\n",
       "      <td>5.4.1.-</td>\n",
       "      <td>cbiC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>CB33.bin.104</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      contig start   end strand       locus_tag  \\\n",
       "0  CB33_1618  1051  1809      -  HIBBKILN_00002   \n",
       "1  CB33_1715   109   261      +  HIBBKILN_00005   \n",
       "2  CB33_1715  1197  2282      +  HIBBKILN_00007   \n",
       "3  CB33_1715  2314  3366      +  HIBBKILN_00008   \n",
       "4  CB33_1715  4105  4746      +  HIBBKILN_00010   \n",
       "\n",
       "                                             product  eC_number  gene db_xref  \\\n",
       "0      rhomboid family intramembrane serine protease        NaN   NaN     NaN   \n",
       "1                    S-adenosylmethionine synthetase    2.5.1.6  metK     NaN   \n",
       "2  putative cobalt-precorrin-6A synthase (deacety...    2.1.1.-   NaN     NaN   \n",
       "3               Uroporphyrin-III C-methyltransferase  2.1.1.107  cbiG     NaN   \n",
       "4                   Cobalt-precorrin-8X methylmutase    5.4.1.-  cbiC     NaN   \n",
       "\n",
       "  path_accession                                          path_name  \\\n",
       "0            NaN                                                NaN   \n",
       "1       PWY-6292  superpathway of L-cysteine biosynthesis (mamma...   \n",
       "2       PWY-6146  <i>Methanobacterium thermoautotrophicum</i> bi...   \n",
       "3       PWY-5196                            factor 430 biosynthesis   \n",
       "4            NaN                                                NaN   \n",
       "\n",
       "   metacyc_target        genome  \n",
       "0           False  CB33.bin.104  \n",
       "1           False  CB33.bin.104  \n",
       "2           False  CB33.bin.104  \n",
       "3           False  CB33.bin.104  \n",
       "4           False  CB33.bin.104  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mc_pathway_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "sexual-silly",
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_presabs = categorical_pivot(mc_pathway_df, 'db_xref', 'genome')\n",
    "\n",
    "kegg_fegenie_annots = '/Volumes/KeithSSD/CBFunctions/data/outputs/gene_annotations2.p'\n",
    "prot_dir = '/Volumes/KeithSSD/CBFunctions/data/ORF_calls'\n",
    "kf_pathway_df = load_kegg_database(kegg_fegenie_annots, qt_idx, prot_dir)\n",
    "\n",
    "target_genomes = set(mc_pathway_df.genome).intersection(set(kf_pathway_df.genome))\n",
    "pathway_df = merge_databases(mc_pathway_df, kf_pathway_df, target_genomes).drop_duplicates()\n",
    "pathway_df['merger_col'] = pathway_df[['genome', 'contig', 'start', 'end']].apply(tuple, axis=1)\n",
    "pathway_df['identifier'] = pathway_df['genome'] + \"_\" + pathway_df['contig'] + \"_\" + pathway_df['start'] + \"_\" + pathway_df['end']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fatal-lucas",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "pathway_df.to_csv(\"~/OneDrive - Johns Hopkins/CB_16S_Shotgun_Manuscript/Data/bin_annotations_merged.txt\", sep=\"\\t\", index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "productive-bottle",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_annotations_into_matrix(pathway_df):\n",
    "    sarNO = ~pathway_df['product'].astype(str).apply(lambda x: 'arcosine' in x)\n",
    "    \n",
    "    annotations = {'psrA': pathway_df['product'].astype(str).str.lower().apply(lambda x: 'polysulfide reductase' in x) | \n",
    "                           (pathway_df['gene'].astype(str).str.startswith('psrA') & ~pathway_df['product'].astype(str).str.lower().str.contains('fatty')),\n",
    "                   'phs': pathway_df.gene.astype(str).str.contains('phs'),\n",
    "                   'fccA': pathway_df.gene.astype(str).str.startswith('fccA') & ~pathway_df['product'].astype(str).str.contains('umarate'),\n",
    "                   'fccB': pathway_df.gene.astype(str).str.startswith('fccB') & ~pathway_df['product'].astype(str).str.contains('umarate'),\n",
    "                   'hydG': pathway_df['product'].astype(str).apply(lambda x: 'ulfhydrogenase' in x) | (pathway_df.query_id.astype(str) == 'K17995'),\n",
    "                   'aprA': pathway_df.gene.astype(str).str.startswith('aprA') | (pathway_df.query_id.astype(str) == 'K00394'),\n",
    "                   'aprB': pathway_df.gene.astype(str).str.startswith('aprB') | (pathway_df.query_id.astype(str) == 'K00395'),\n",
    "                   'asrA': pathway_df.gene.astype(str).str.startswith('asrA') | (pathway_df.query_id.astype(str) == 'K16950'),\n",
    "                   'asrB': pathway_df.gene.astype(str).str.startswith('asrB') | (pathway_df.query_id.astype(str) == 'K16951'),\n",
    "                   'dsrE/tusD': pathway_df.query_id.isin(['K07235']) | pathway_df.gene.astype(str).str.startswith(\"dsrE\") | pathway_df['product'].astype(str).str.lower().str.contains('tusd') | pathway_df.gene.astype(str).str.startswith(\"tusD\"),\n",
    "                   'dsrF/tusC': pathway_df.query_id.isin(['K07236']) | pathway_df.gene.astype(str).str.startswith(\"dsrF\") | pathway_df['product'].astype(str).str.lower().str.contains('tusc') | pathway_df.gene.astype(str).str.startswith(\"tusC\"),\n",
    "                   'dsrH/tusB': pathway_df.query_id.isin(['K07237']) | pathway_df.gene.astype(str).str.startswith(\"dsrH\") | pathway_df['product'].astype(str).str.lower().str.contains('tusb') | pathway_df.gene.astype(str).str.startswith(\"tusB\"),\n",
    "                   'tsdA': (pathway_df.query_id.astype(str) == 'K19713'),\n",
    "                   'sat/cysN/cysD/cysC/cysNC/cysH': pathway_df.query_id.isin(['K00958', 'K00860', 'K00955', 'K00956', 'K00957','K00390']) &  ~pathway_df['product'].astype(str).apply(lambda x: 'alanine' in x) & ~pathway_df['product'].astype(str).apply(lambda x: 'elongation' in x),\n",
    "                   'soeA': pathway_df.gene.astype(str).str.startswith('soe') & (pathway_df.query_id.astype(str) == 'K21307'),\n",
    "                   'soeB': pathway_df.gene.astype(str).str.startswith('soe') & (pathway_df.query_id.astype(str) == 'K21308'),\n",
    "                   'soeC': pathway_df.gene.astype(str).str.startswith('soe') & (pathway_df.query_id.astype(str) == 'K21309'),\n",
    "                   'sqr': (pathway_df.query_id.astype(str) == 'K17218'),\n",
    "                   'soxL': pathway_df.gene.astype(str).str.startswith('soxL'),\n",
    "                   'msrB': pathway_df.gene.astype(str).str.startswith('msrB'),\n",
    "                   'msrA': pathway_df.gene.astype(str).str.startswith('msrA'),\n",
    "                   'dsrC': (pathway_df.gene.astype(str).str.startswith('dsrC') | pathway_df.gene.astype(str).str.startswith('tusE') | (pathway_df['product'] == 'sulfite reductase subunit gamma') | pathway_df['product'].astype(str).str.lower().str.contains('tuse')),\n",
    "                   'soxY': ((pathway_df.query_id.astype(str) == 'K17226') | pathway_df['gene'].astype(str).str.startswith('soxY') | pathway_df['product'].astype(str).str.lower().str.contains('soxy')) & sarNO,\n",
    "                   'soxZ': ((pathway_df.query_id.astype(str) == 'K17227') | pathway_df['gene'].astype(str).str.startswith('soxZ') | pathway_df['product'].astype(str).str.lower().str.contains('soxz')) & sarNO,\n",
    "                   \"dsrA\": (pathway_df.query_id.astype(str) == 'K11181') | pathway_df['gene'].astype(str).str.startswith('dsrB') | pathway_df['gene'].astype(str).str.startswith('dsvB'),\n",
    "                   \"dsrB\": (pathway_df.query_id.astype(str) == 'K11180') | pathway_df['gene'].astype(str).str.startswith('dsrA') | pathway_df['gene'].astype(str).str.startswith('dsvA'),\n",
    "                   \"dsrL\": (pathway_df['gene'].astype(str).str.startswith('dsrL') | pathway_df['product'].astype(str).str.lower().str.contains('dsrL')),\n",
    "                   'soxA': ((pathway_df.query_id.astype(str) == 'K17222') | pathway_df['gene'].astype(str).str.startswith('soxA') | pathway_df['product'].astype(str).str.lower().str.contains('soxa')) & sarNO,\n",
    "                   'soxX': ((pathway_df.query_id.astype(str) == 'K17223') | pathway_df['gene'].astype(str).str.startswith('soxX') | pathway_df['product'].astype(str).str.lower().str.contains('soxx')) & sarNO,\n",
    "                   'soxB': ((pathway_df.query_id.astype(str) == 'K17224') | pathway_df['gene'].astype(str).str.startswith('soxB') | pathway_df['product'].astype(str).str.lower().str.contains('soxb')) & sarNO,\n",
    "                   'soxD': ((pathway_df.query_id.astype(str) == 'K22622') | pathway_df['gene'].astype(str).str.startswith('soxD') | pathway_df['product'].astype(str).str.lower().str.contains('soxd')) & sarNO,\n",
    "                   'soxC': ((pathway_df.query_id.astype(str) == 'K17225') | pathway_df['gene'].astype(str).str.startswith('soxC') | pathway_df['product'].astype(str).str.lower().str.contains('soxc')) & sarNO,\n",
    "                   'rdhA/thtR/sseB': ((pathway_df.query_id.astype(str) == 'K01011') | pathway_df['gene'].astype(str).str.startswith('rdhA')),\n",
    "                   'sir/cysJI': (pathway_df['gene'].astype(str).str.startswith('sir_') | (pathway_df['gene'].astype(str) == 'sir') | pathway_df['gene'].astype(str).str.startswith('cysJ') | pathway_df['gene'].astype(str).str.startswith('cysI')) | pathway_df.query_id.isin(['K00381', 'K00380', 'K00392']),\n",
    "                   'rbcS': (pathway_df.query_id.astype(str) == 'K01602') | (pathway_df.gene.astype(str).apply(lambda x: 'ibulose' in x) & pathway_df.gene.astype(str).apply(lambda x: 'isphosphate' in x) & pathway_df.gene.astype(str).apply(lambda x: 'small' in x.lower())),\n",
    "                   'rbcL': (pathway_df.query_id.astype(str) == 'K01601') | (pathway_df.gene.astype(str).apply(lambda x: 'ibulose' in x) & pathway_df.gene.astype(str).apply(lambda x: 'isphosphate' in x) & pathway_df.gene.astype(str).apply(lambda x: 'large' in x.lower()))}\n",
    "    \n",
    "    complexes = {'fccAB': (\"fccA\", \"fccB\",), #> 1\n",
    "                 'aprAB': (\"aprA\", \"aprB\",), #> 1\n",
    "                 'msrAB': (\"msrA\", \"msrB\",), #> 1\n",
    "                 'asrAB': (\"asrA\", \"asrB\",), #> 1\n",
    "                 'dsrEFH/tusBCD': (\"dsrE/tusD\", \"dsrF/tusC\", \"dsrH/tusB\",), # > 1\n",
    "                 'soeABC': (\"soeA\", \"soeB\", \"soeC\",), # > 1\n",
    "                 'soxYZ': (\"soxY\", \"soxZ\",),\n",
    "                 'dsrAB': (\"dsrA\", \"dsrB\",),\n",
    "                 'soxAX': (\"soxA\", \"soxX\",),\n",
    "                 'soxCD': (\"soxD\", \"soxC\",),\n",
    "                 'rbcSL': (\"rbcS\", \"rbcL\",)}\n",
    "    \n",
    "    quinol_oxidases_names = ['appB', 'appC', 'cioB', 'cydA', 'cydB', 'cydX', 'cyoA', 'cyoB', 'cyoC', 'cyoD', 'qoxC']\n",
    "    quinol_oxidases_ids = ['K00426', 'K00425', 'K02299']\n",
    "    high_o2_oxidases = ['K02276', 'K02298', 'K02297', 'K02300']\n",
    "\n",
    "    full_pathways = {'taurine_degradation': (pathway_df.path_name.astype(str).str.contains('taurine degradation'), 1, 'locus_tag'),\n",
    "                     'coxLMS/cutLMS': (pathway_df['product'].astype(str).str.contains('arbon monoxide dehydrogenase'), 1, 'locus_tag'),\n",
    "                     'appBC/cydABX/qoxC':   (pathway_df.query_id.isin(quinol_oxidases_ids) | \n",
    "                                             pathway_df.gene.astype(str).str.startswith('appB') |  # anoxic conditions\n",
    "                                             pathway_df.gene.astype(str).str.startswith('appC') |  # anoxic conditions\n",
    "                                             pathway_df.gene.astype(str).str.startswith('cioB') | # microaerobic\n",
    "                                             pathway_df.gene.astype(str).str.startswith('cydA') | # microaerobic\n",
    "                                             pathway_df.gene.astype(str).str.startswith('cydB') | # microaerobic\n",
    "                                             pathway_df.gene.astype(str).str.startswith('cydX') | # microaerobic\n",
    "                                             pathway_df.gene.astype(str).str.startswith('qoxC')  & \n",
    "                                             (pathway_df.query_id != \"K02033\"), 1, 'locus_tag'),\n",
    "                     'accABCD/pccAB': (pathway_df.gene.astype(str).str.startswith('accA') | \n",
    "                                       pathway_df.gene.astype(str).str.startswith('accB') | \n",
    "                                       pathway_df.gene.astype(str).str.startswith('accC') | \n",
    "                                       pathway_df.gene.astype(str).str.startswith('accD') | \n",
    "                                       pathway_df.gene.astype(str).str.startswith('pccA') | \n",
    "                                       pathway_df.gene.astype(str).str.startswith('pccB') , 1, 'locus_tag'),\n",
    "                     \"napAB+narIJ\": ((pathway_df.query_id.isin(NITROGEN_CYCLE['nitrate reductase (napAB/narIJ)']) & pathway_df.query_id.notnull()), 1, 'query_id'),\n",
    "                     'amoABC': ((pathway_df.query_id.isin(NITROGEN_CYCLE['ammonia oxidation (amoABC)']) & pathway_df.query_id.notnull()), 1, 'query_id'),\n",
    "                     'nirK/nirS+norBC': (pathway_df.query_id.isin({'K15864', 'K04561', 'K00368', 'K02305',}) |\n",
    "                                         pathway_df.gene.astype(str).str.startswith('norB') | \n",
    "                                         pathway_df.gene.astype(str).str.startswith('norC') |\n",
    "                                         pathway_df.gene.astype(str).str.startswith('nirK') |\n",
    "                                         pathway_df.gene.astype(str).str.startswith('nirS'), 1, 'query_id'),\n",
    "                     'nosZ': ((pathway_df.query_id == 'K00376') | pathway_df.gene.astype(str).str.startswith('nosZ'), 1, 'query_id'),\n",
    "                     'nirABD': (pathway_df.query_id.isin({'K00363', 'K00362', 'K00366'}) |\n",
    "                                      pathway_df.gene.astype(str).str.startswith('nirA') |\n",
    "                                      pathway_df.gene.astype(str).str.startswith('nirB') |\n",
    "                                      pathway_df.gene.astype(str).str.startswith('nirD'), 1, 'query_id'),\n",
    "                     'nrfAH': (pathway_df.query_id.isin({'K15876', 'K03385'}) | pathway_df.gene.astype(str).str.startswith('nrfH') |\n",
    "                               pathway_df.gene.astype(str).str.startswith('nrfA'), 0, 'query_id'),\n",
    "                     'nifKHDW/vnfA': (pathway_df['product'].astype(str).str.lower().str.contains('nitrogenase'), 1, 'query_id'),\n",
    "                     'coxABC/ctaCDE/cyoABCD': ((pathway_df['product'].astype(str).str.lower().str.contains('cytochrome c oxidase') & \n",
    "                                               (pathway_df['gene'].astype(str).str.startswith('cta') | pathway_df['gene'].astype(str).str.startswith('cox'))) |\n",
    "                                              (pathway_df.gene.astype(str).str.startswith('cyoA') | \n",
    "                                               pathway_df.gene.astype(str).str.startswith('cyoB') | \n",
    "                                               pathway_df.gene.astype(str).str.startswith('cyoC') | \n",
    "                                               pathway_df.gene.astype(str).str.startswith('cyoD'))  | \n",
    "                                               pathway_df.query_id.isin(high_o2_oxidases), 1, 'locus_tag'),\n",
    "                     'katGBE/sodABC/dfx':(pathway_df.gene.astype(str).str.startswith('grxC') | \n",
    "                                      pathway_df.gene.astype(str).str.startswith('dfx') | \n",
    "                                      pathway_df['product'].astype(str).str.lower().str.contains('alkylhydroperoxidase') | \n",
    "                                      pathway_df['product'].astype(str).str.lower().str.contains('oxide dismutase') | \n",
    "                                      pathway_df['product'].astype(str).str.lower().str.contains('oxid dismutase') | \n",
    "                                      pathway_df['product'].astype(str).str.lower().str.contains('catalase'), 1, 'locus_tag'),\n",
    "                     'psaABCDEFIJKL':(pathway_df.gene.astype(str).str.startswith('psa'), 9, 'locus_tag'),\n",
    "                     'petABCDGLMN':(pathway_df.gene.astype(str).str.startswith('pet'), 10, 'locus_tag'),\n",
    "                     'pufABCLMX':(pathway_df.gene.astype(str).str.startswith('puf'), 1, 'locus_tag'),\n",
    "                     'psbA-Z':(pathway_df.gene.astype(str).str.startswith('psb'), 3, 'locus_tag'),\n",
    "                     'hoxS/hndABC': (pathway_df.gene.astype(str).str.startswith('hox') | \n",
    "                                     pathway_df.gene.astype(str).str.startswith('hnd'), 1, 'locus_tag'),\n",
    "                     'hupLBS/hoxZKG': (pathway_df.gene.astype(str).str.startswith('hup'), 1, 'locus_tag'),\n",
    "                     'cpeABCDERSTYZ': (pathway_df.gene.astype(str).str.startswith('cpe'), 1, 'locus_tag'),\n",
    "                     'fdnG/fdhF/fdhAB/fdoGHI/fdsAB': (pathway_df['product'].astype(str).str.lower().str.contains('formate dehydrogenase') & \n",
    "                                        ~pathway_df['product'].astype(str).str.lower().str.contains('mitochondrial'), 1, 'locus_tag'),\n",
    "                     'hao': (pathway_df.gene.astype(str).str.startswith('hao'), 0, 'locus_tag')}\n",
    "    \n",
    "    in_complex = set()\n",
    "    is_complex = set()\n",
    "    genomes_with = {}\n",
    "    genes_within = {}\n",
    "    for k, v in complexes.items():\n",
    "        is_complex.add(k)\n",
    "        in_complex.update([i for i in v])\n",
    "        \n",
    "        if len(v) == 2:\n",
    "            annotations[k] = annotations[v[0]] | annotations[v[1]]\n",
    "        \n",
    "            have_complex = np.unique(list(pathway_df[annotations[v[0]]]['genome'].unique()) + \\\n",
    "                                      list(pathway_df[annotations[v[1]]]['genome'].unique()), return_counts=True)\n",
    "        elif len(v) == 3:\n",
    "            annotations[k] = annotations[v[0]] | annotations[v[1]] | annotations[v[2]]\n",
    "            have_complex = np.unique(list(pathway_df[annotations[v[0]]]['genome'].unique()) + \\\n",
    "                                      list(pathway_df[annotations[v[1]]]['genome'].unique()) + \\\n",
    "                                      list(pathway_df[annotations[v[2]]]['genome'].unique()), return_counts=True)\n",
    "        \n",
    "        genes_within[k] = set(pathway_df.loc[annotations[k], 'identifier'])\n",
    "        genomes_with[k] = set(have_complex[0][have_complex[1] > 1])\n",
    "    \n",
    "    for k, v in annotations.items():\n",
    "        if (k not in in_complex) and (k not in is_complex):\n",
    "            genomes_with[k] = set(pathway_df[v]['genome'])\n",
    "            genes_within[k] = set(pathway_df.loc[annotations[k], 'identifier'])\n",
    "    \n",
    "    for fp, (bool_i, cutoff_i, query_col) in full_pathways.items():\n",
    "        agg_df = pathway_df.loc[bool_i, ['genome', query_col]].groupby('genome').agg(lambda x: len(set(x.values)))[query_col]\n",
    "        genomes_with[fp] = set(agg_df.index[agg_df > cutoff_i])\n",
    "        genes_within[fp] = set(pathway_df.loc[bool_i & pathway_df.genome.isin(genomes_with[fp]), 'identifier'])\n",
    "    \n",
    "    new_annotations = {}\n",
    "    for key, val in genomes_with.items():\n",
    "        new_annotations[key] = {k:True if k in val else False for k in pathway_df.genome.unique()}\n",
    "    \n",
    "    annot_mat = pd.DataFrame(new_annotations)\n",
    "    return annot_mat, genes_within "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "joined-enough",
   "metadata": {},
   "outputs": [],
   "source": [
    "annot_mat, gene_sets = parse_annotations_into_matrix(pathway_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "aerial-venice",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>product</th>\n",
       "      <th>gene</th>\n",
       "      <th>query_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>176907</th>\n",
       "      <td>Cytochrome c oxidase polypeptide I+III</td>\n",
       "      <td>caaA</td>\n",
       "      <td>K02276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>218780</th>\n",
       "      <td>Cytochrome c oxidase polypeptide I+III</td>\n",
       "      <td>caaA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274237</th>\n",
       "      <td>Cytochrome c oxidase polypeptide I+III</td>\n",
       "      <td>caaA_1</td>\n",
       "      <td>K02276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285840</th>\n",
       "      <td>Cytochrome c oxidase polypeptide I+III</td>\n",
       "      <td>caaA_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68879</th>\n",
       "      <td>cytochrome c oxidase subunit I</td>\n",
       "      <td>coxA</td>\n",
       "      <td>K02274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36004</th>\n",
       "      <td>Cytochrome bo(3) ubiquinol oxidase subunit 3</td>\n",
       "      <td>cyoC</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339142</th>\n",
       "      <td>cytochrome o ubiquinol oxidase subunit IV</td>\n",
       "      <td>cyoD_1</td>\n",
       "      <td>K02300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>343631</th>\n",
       "      <td>cytochrome o ubiquinol oxidase subunit IV</td>\n",
       "      <td>cyoD_2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147324</th>\n",
       "      <td>Cytochrome c6</td>\n",
       "      <td>petJ</td>\n",
       "      <td>K02275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293551</th>\n",
       "      <td>Quinol oxidase subunit 3</td>\n",
       "      <td>qoxC</td>\n",
       "      <td>K02276</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>137 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             product    gene query_id\n",
       "176907        Cytochrome c oxidase polypeptide I+III    caaA   K02276\n",
       "218780        Cytochrome c oxidase polypeptide I+III    caaA      NaN\n",
       "274237        Cytochrome c oxidase polypeptide I+III  caaA_1   K02276\n",
       "285840        Cytochrome c oxidase polypeptide I+III  caaA_2      NaN\n",
       "68879                 cytochrome c oxidase subunit I    coxA   K02274\n",
       "...                                              ...     ...      ...\n",
       "36004   Cytochrome bo(3) ubiquinol oxidase subunit 3    cyoC      NaN\n",
       "339142     cytochrome o ubiquinol oxidase subunit IV  cyoD_1   K02300\n",
       "343631     cytochrome o ubiquinol oxidase subunit IV  cyoD_2      NaN\n",
       "147324                                 Cytochrome c6    petJ   K02275\n",
       "293551                      Quinol oxidase subunit 3    qoxC   K02276\n",
       "\n",
       "[137 rows x 3 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pathway_df.loc[pathway_df['identifier'].isin(gene_sets['coxABC/ctaCDE/cyoABCD']) & pathway_df['gene'].notnull(), \n",
    "               ['product', 'gene', 'query_id']].drop_duplicates().sort_values(by='gene')\n",
    "\n",
    "\n",
    "#pathway_df.loc[pathway_df['identifier'].isin(gene_sets['appBC/cydABX/qoxC']) & pathway_df['gene'].notnull(), \n",
    "#               ['product', 'gene', 'query_id']].drop_duplicates().sort_values(by='gene')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "incredible-reviewer",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote 1120 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.104.genes.fa\n",
      "Wrote 1558 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.111.genes.fa\n",
      "Wrote 2163 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.118.genes.fa\n",
      "Wrote 2797 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.133.genes.fa\n",
      "Wrote 2530 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.142.genes.fa\n",
      "Wrote 1573 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.146.genes.fa\n",
      "Wrote 2560 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.149.genes.fa\n",
      "Wrote 2068 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.167.genes.fa\n",
      "Wrote 1848 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.168.genes.fa\n",
      "Wrote 2181 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.169.genes.fa\n",
      "Wrote 1280 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.172.genes.fa\n",
      "Wrote 2731 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.174.genes.fa\n",
      "Wrote 3098 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.175.genes.fa\n",
      "Wrote 2259 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.179.genes.fa\n",
      "Wrote 2193 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.182.genes.fa\n",
      "Wrote 2463 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.185.genes.fa\n",
      "Wrote 3497 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.186.genes.fa\n",
      "Wrote 1980 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.190.genes.fa\n",
      "Wrote 1072 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.191.genes.fa\n",
      "Wrote 1484 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.194.genes.fa\n",
      "Wrote 3178 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.196.genes.fa\n",
      "Wrote 1071 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.2.genes.fa\n",
      "Wrote 2172 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.201.genes.fa\n",
      "Wrote 2616 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.202.genes.fa\n",
      "Wrote 1776 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.203.genes.fa\n",
      "Wrote 1608 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.207.genes.fa\n",
      "Wrote 1892 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.212.genes.fa\n",
      "Wrote 2371 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.216.genes.fa\n",
      "Wrote 2389 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.218.genes.fa\n",
      "Wrote 3298 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.219.genes.fa\n",
      "Wrote 2826 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.222.genes.fa\n",
      "Wrote 2053 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.227.genes.fa\n",
      "Wrote 2262 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.229.genes.fa\n",
      "Wrote 1841 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.233.genes.fa\n",
      "Wrote 3151 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.235.genes.fa\n",
      "Wrote 2700 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.236.genes.fa\n",
      "Wrote 1515 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.32.genes.fa\n",
      "Wrote 1393 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.34.genes.fa\n",
      "Wrote 2245 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.56.genes.fa\n",
      "Wrote 1482 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.61.genes.fa\n",
      "Wrote 1556 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.63.genes.fa\n",
      "Wrote 1776 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.82.genes.fa\n",
      "Wrote 1062 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.84.genes.fa\n",
      "Wrote 1225 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.85.genes.fa\n",
      "Wrote 1303 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.86.genes.fa\n",
      "Wrote 2302 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.89.genes.fa\n",
      "Wrote 2193 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.90.genes.fa\n",
      "Wrote 2763 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CB33.bin.95.genes.fa\n",
      "Wrote 1278 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.111.genes.fa\n",
      "Wrote 2932 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.113.genes.fa\n",
      "Wrote 1960 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.118.genes.fa\n",
      "Wrote 1740 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.142.genes.fa\n",
      "Wrote 3182 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.174.genes.fa\n",
      "Wrote 1326 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.188.genes.fa\n",
      "Wrote 1816 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.189.genes.fa\n",
      "Wrote 1574 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.191.genes.fa\n",
      "Wrote 2283 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.196.genes.fa\n",
      "Wrote 2393 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.222.genes.fa\n",
      "Wrote 1861 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.223.genes.fa\n",
      "Wrote 968 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.246.genes.fa\n",
      "Wrote 1172 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.25.genes.fa\n",
      "Wrote 1914 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.252.genes.fa\n",
      "Wrote 6450 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.26.genes.fa\n",
      "Wrote 1193 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.263.genes.fa\n",
      "Wrote 3270 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.287.genes.fa\n",
      "Wrote 2525 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.322.genes.fa\n",
      "Wrote 2076 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.327.genes.fa\n",
      "Wrote 2359 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.331.genes.fa\n",
      "Wrote 3263 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.333.genes.fa\n",
      "Wrote 2598 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.344.genes.fa\n",
      "Wrote 1532 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.346.genes.fa\n",
      "Wrote 1909 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.349.genes.fa\n",
      "Wrote 1241 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.355.genes.fa\n",
      "Wrote 2807 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.358.genes.fa\n",
      "Wrote 2164 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.359.genes.fa\n",
      "Wrote 1826 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.368.genes.fa\n",
      "Wrote 2547 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.370.genes.fa\n",
      "Wrote 2958 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.374.genes.fa\n",
      "Wrote 1332 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.375.genes.fa\n",
      "Wrote 2389 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.377.genes.fa\n",
      "Wrote 1651 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.381.genes.fa\n",
      "Wrote 1877 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.384.genes.fa\n",
      "Wrote 2133 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.385.genes.fa\n",
      "Wrote 2489 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.388.genes.fa\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wrote 2771 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.389.genes.fa\n",
      "Wrote 1507 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.390.genes.fa\n",
      "Wrote 2078 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.394.genes.fa\n",
      "Wrote 2227 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.395.genes.fa\n",
      "Wrote 1119 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.40.genes.fa\n",
      "Wrote 3086 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.405.genes.fa\n",
      "Wrote 2390 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.408.genes.fa\n",
      "Wrote 2302 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.6.genes.fa\n",
      "Wrote 2575 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.61.genes.fa\n",
      "Wrote 1694 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.63.genes.fa\n",
      "Wrote 2190 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.84.genes.fa\n",
      "Wrote 1034 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.94.genes.fa\n",
      "Wrote 2189 (100.0%) sequences to /Volumes/KeithSSD/SulFox/data/pulled_genes/CBrest.bin.96.genes.fa\n"
     ]
    }
   ],
   "source": [
    "seq_map = pull_genes(pathway_df, '/Volumes/KeithSSD/CBFunctions/data/all_bins', \n",
    "                                 '/Volumes/KeithSSD/SulFox/data/pulled_genes', '.fa')\n",
    "seq_map = {\"_\".join(k):v for k, v in seq_map.items()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "fitting-extreme",
   "metadata": {},
   "outputs": [],
   "source": [
    "# genes used to normalize\n",
    "housekeepers = [\"proC\", \"recA\", \"rpoD\", \"rho\", \"glyA\", \"tpiA\", \"recF\"]\n",
    "# \n",
    "gene_code_lookup = {}\n",
    "gene_id_columns = ['genome', 'contig', 'start', 'end']\n",
    "id_dicter = lambda x: {\"_\".join(list(x)[:-1]): list(x)[-1]}\n",
    "gene_code_lookup_chunks = list(pathway_df[gene_id_columns + ['gene']].dropna().apply(id_dicter, axis=1))\n",
    "for i in gene_code_lookup_chunks:\n",
    "    gene_code_lookup.update(i)\n",
    "\n",
    "\n",
    "hk_per_bin = defaultdict(set)\n",
    "for h in housekeepers:\n",
    "    hbool = (pathway_df.gene == h)\n",
    "    gene_ids = pathway_df.loc[hbool, gene_id_columns].apply(lambda x: \"_\".join(list(x)), axis=1).to_frame()\n",
    "    gene_ids = (pathway_df.loc[hbool, ['genome']]).join(gene_ids)\n",
    "    gene_ids = gene_ids.groupby('genome').agg(set)[0]\n",
    "    for g in gene_ids.index:\n",
    "        hk_per_bin[g].update(gene_ids[g])\n",
    "\n",
    "for g in hk_per_bin:\n",
    "    assert len(hk_per_bin[g]) > 0\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "interpreted-pollution",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "92 bins' transcripts were mapped\n",
      "38 samples probed\n",
      "[934, 2849]\n"
     ]
    }
   ],
   "source": [
    "salmon_folder = '/Volumes/KeithSSD/SulFox/data/salmon_out4'\n",
    "\n",
    "avail_samples = defaultdict(list)\n",
    "avail_bins = defaultdict(list)\n",
    "salmon_files = defaultdict(list)\n",
    "passfail = [0, 0]\n",
    "for i in os.listdir(salmon_folder):\n",
    "    bincheck = [1 for b in bins_desired if f\"{b}.quant.sf\" in i]\n",
    "    if sum(bincheck) > 0:\n",
    "        sample_i = i.split(\"_CB\")[0]\n",
    "        bin_i = i.replace(f'{sample_i}_', '').replace(\".quant.sf\", \"\")\n",
    "        file_i = os.path.join(salmon_folder, i)\n",
    "        if sample_i.replace(\"_pass\", \"\") in hewson_translator:\n",
    "            sample_i = hewson_translator[sample_i.replace(\"_pass\", \"\")]\n",
    "        df = pd.read_csv(file_i, sep=\"\\t\", index_col=0).loc[:, ['TPM']].rename(columns={'TPM':sample_i})\n",
    "        if (df[sample_i] > 0).sum() > int(df.shape[0]*0.1):\n",
    "            salmon_files[bin_i].append(df)\n",
    "            passfail[0] += 1\n",
    "            avail_samples[sample_i].append(bin_i)\n",
    "            avail_bins[bin_i].append(sample_i)\n",
    "        else:\n",
    "            passfail[1] += 1\n",
    "            salmon_files[bin_i].append(df*0)\n",
    "\n",
    "print(len(avail_bins), \"bins' transcripts were mapped\")\n",
    "print(len(avail_samples), 'samples probed')\n",
    "print(passfail)\n",
    "\n",
    "salmon_dfs = {b:pd.concat(v, axis=1, verify_integrity=True) for b, v in salmon_files.items()}\n",
    "salmon_df = pd.concat(salmon_dfs.values(), verify_integrity=True)\n",
    "salmon_df['merger_col'] = pd.Series({i:tuple(i.split(\"_\")) for i in salmon_df.index})\n",
    "salmon_df['gene_name'] = pd.Series({i:gene_code_lookup.get(i) for i in salmon_df.index})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "afraid-bobby",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlYAAAHSCAYAAAAubIVMAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAV/klEQVR4nO3df4zkd33f8dc7PqBpQcXUZ8u1Tc9FThtTFZNeXVTSioS2dvjHRgrV0YpYEZJT1VREyh8B/iipKktUapK2aknkAMKVElyrEHArmtZ1k1KUgDkjY/wjLldM7Yst+4C0kFRy5ePdP3bcLGbPO3f7nt253cdDGs3Md77fmc/uZ2fued/5Vd0dAAB27vv2egAAAPuFsAIAGCKsAACGCCsAgCHCCgBgiLACABhyaK8HkCQXXXRRHzlyZK+HAQCwrfvuu+/r3X14q8vWIqyOHDmS48eP7/UwAAC2VVX/80yXeSoQAGCIsAIAGCKsAACGCCsAgCHCCgBgiLACABgirAAAhggrAIAhwgoAYIiwAgAYIqwAAIYIKwCAIcIKAGCIsAIAGCKsAACGCCsAgCHCCgBgiLACABgirAAAhhza6wHsphtv/Ox3nf/kJ394j0YCAOxH9lgBAAwRVgAAQ4QVAMAQYQUAMERYAQAMEVYAAEOEFQDAEGEFADBEWAEADBFWAABDhBUAwBBhBQAwRFgBAAzZNqyq6o9V1b1V9aWqeqiq/tFi+auq6u6q+sri+MJN27y3qk5U1aNVdd0qfwAAgHWxzB6rZ5P8aHe/Lsk1Sa6vqjckeU+Se7r7qiT3LM6nqq5OcizJa5Ncn+SDVXXBCsYOALBWtg2r3vAHi7MvWRw6yQ1Jbl8svz3JjYvTNyS5o7uf7e7HkpxIcu3koAEA1tFSr7Gqqguq6v4kzyS5u7s/n+SS7n4qSRbHFy9WvyzJE5s2P7lYBgCwry0VVt19uruvSXJ5kmur6i+8yOq11VV8z0pVN1fV8ao6furUqaUGCwCwzs7qXYHd/b+S/FY2Xjv1dFVdmiSL42cWq51McsWmzS5P8uQW13Vbdx/t7qOHDx8++5EDAKyZZd4VeLiqXrk4/f1J/kaS301yV5KbFqvdlORTi9N3JTlWVS+rqiuTXJXk3uFxAwCsnUNLrHNpktsX7+z7viR3dve/r6rfSXJnVb0zyeNJ3pYk3f1QVd2Z5OEkzyW5pbtPr2b4AADrY9uw6u4Hkrx+i+XfSPLmM2xza5Jbdzw6AIDziE9eBwAYIqwAAIYIKwCAIcIKAGCIsAIAGCKsAACGCCsAgCHCCgBgiLACABgirAAAhggrAIAhwgoAYIiwAgAYIqwAAIYIKwCAIcIKAGCIsAIAGCKsAACGCCsAgCHCCgBgiLACABgirAAAhggrAIAhwgoAYIiwAgAYIqwAAIYIKwCAIcIKAGCIsAIAGCKsAACGCCsAgCHCCgBgiLACABgirAAAhggrAIAhwgoAYIiwAgAYIqwAAIYIKwCAIcIKAGCIsAIAGCKsAACGCCsAgCHCCgBgiLACABgirAAAhggrAIAhwgoAYIiwAgAYIqwAAIYIKwCAIcIKAGCIsAIAGCKsAACGCCsAgCHCCgBgiLACABiybVhV1RVV9ZtV9UhVPVRV714s/7mq+r2qun9xeMumbd5bVSeq6tGqum6VPwAAwLo4tMQ6zyX5me7+YlW9Isl9VXX34rJf7O5/unnlqro6ybEkr03yp5P856r6ge4+PTlwAIB1s+0eq+5+qru/uDj97SSPJLnsRTa5Ickd3f1sdz+W5ESSaycGCwCwzs7qNVZVdSTJ65N8frHoXVX1QFV9pKouXCy7LMkTmzY7mRcPMQCAfWHpsKqqlyf5eJKf7u5vJfmlJK9Jck2Sp5L8/POrbrF5b3F9N1fV8ao6furUqbMdNwDA2lkqrKrqJdmIql/t7k8kSXc/3d2nu/s7SX4lf/R038kkV2za/PIkT77wOrv7tu4+2t1HDx8+vJOfAQBgLSzzrsBK8uEkj3T3L2xafumm1d6a5MHF6buSHKuql1XVlUmuSnLv3JABANbTMu8KfGOSdyT5clXdv1j2viRvr6prsvE039eS/FSSdPdDVXVnkoez8Y7CW7wjEAA4CLYNq+7+bLZ+3dSnX2SbW5PcuoNxAQCcd3zyOgDAEGEFADBEWAEADBFWAABDhBUAwBBhBQAwRFgBAAwRVgAAQ4QVAMAQYQUAMERYAQAMEVYAAEOEFQDAEGEFADBEWAEADBFWAABDhBUAwBBhBQAwRFgBAAwRVgAAQ4QVAMAQYQUAMERYAQAMEVYAAEOEFQDAEGEFADBEWAEADBFWAABDhBUAwBBhBQAwRFgBAAwRVgAAQ4QVAMAQYQUAMERYAQAMEVYAAEOEFQDAEGEFADBEWAEADBFWAABDhBUAwBBhBQAwRFgBAAwRVgAAQ4QVAMAQYQUAMERYAQAMEVYAAEOEFQDAEGEFADBEWAEADBFWAABDhBUAwBBhBQAwRFgBAAwRVgAAQ4QVAMCQbcOqqq6oqt+sqkeq6qGqevdi+auq6u6q+sri+MJN27y3qk5U1aNVdd0qfwAAgHWxzB6r55L8THf/YJI3JLmlqq5O8p4k93T3VUnuWZzP4rJjSV6b5PokH6yqC1YxeACAdbJtWHX3U939xcXpbyd5JMllSW5IcvtitduT3Lg4fUOSO7r72e5+LMmJJNcOjxsAYO2c1WusqupIktcn+XySS7r7qWQjvpJcvFjtsiRPbNrs5GIZAMC+tnRYVdXLk3w8yU9397debNUtlvUW13dzVR2vquOnTp1adhgAAGtrqbCqqpdkI6p+tbs/sVj8dFVdurj80iTPLJafTHLFps0vT/LkC6+zu2/r7qPdffTw4cPnOn4AgLWxzLsCK8mHkzzS3b+w6aK7kty0OH1Tkk9tWn6sql5WVVcmuSrJvXNDBgBYT4eWWOeNSd6R5MtVdf9i2fuSfCDJnVX1ziSPJ3lbknT3Q1V1Z5KHs/GOwlu6+/T0wAEA1s22YdXdn83Wr5tKkjefYZtbk9y6g3EBAJx3fPI6AMAQYQUAMERYAQAMEVYAAEOEFQDAEGEFADBEWAEADBFWAABDhBUAwBBhBQAwRFgBAAwRVgAAQ4QVAMAQYQUAMERYAQAMEVYAAEOEFQDAEGEFADBEWAEADBFWAABDhBUAwBBhBQAwRFgBAAwRVgAAQ4QVAMAQYQUAMERYAQAMEVYAAEOEFQDAEGEFADBEWAEADBFWAABDhBUAwBBhBQAwRFgBAAwRVgAAQ4QVAMAQYQUAMERYAQAMEVYAAEOEFQDAEGEFADBEWAEADBFWAABDhBUAwBBhBQAwRFgBAAwRVgAAQ4QVAMAQYQUAMERYAQAMEVYAAEOEFQDAEGEFADBEWAEADBFWAABDhBUAwJBtw6qqPlJVz1TVg5uW/VxV/V5V3b84vGXTZe+tqhNV9WhVXbeqgQMArJtl9lh9NMn1Wyz/xe6+ZnH4dJJU1dVJjiV57WKbD1bVBVODBQBYZ9uGVXd/Jsk3l7y+G5Lc0d3PdvdjSU4kuXYH4wMAOG/s5DVW76qqBxZPFV64WHZZkic2rXNysQwAYN8717D6pSSvSXJNkqeS/PxieW2xbm91BVV1c1Udr6rjp06dOsdhAACsj3MKq+5+urtPd/d3kvxK/ujpvpNJrti06uVJnjzDddzW3Ue7++jhw4fPZRgAAGvlnMKqqi7ddPatSZ5/x+BdSY5V1cuq6sokVyW5d2dDBAA4PxzaboWq+liSNyW5qKpOJnl/kjdV1TXZeJrva0l+Kkm6+6GqujPJw0meS3JLd59eycgBANbMtmHV3W/fYvGHX2T9W5PcupNBAQCcj3zyOgDAEGEFADBEWAEADBFWAABDhBUAwBBhBQAwRFgBAAwRVgAAQ4QVAMAQYQUAMERYAQAMEVYAAEOEFQDAEGEFADBEWAEADBFWAABDhBUAwBBhBQAwRFgBAAwRVgAAQ4QVAMAQYQUAMERYAQAMEVYAAEOEFQDAEGEFADBEWAEADBFWAABDhBUAwBBhBQAwRFgBAAwRVgAAQ4QVAMAQYQUAMERYAQAMEVYAAEOEFQDAEGEFADBEWAEADBFWAABDhBUAwBBhBQAwRFgBAAwRVgAAQ4QVAMAQYQUAMERYAQAMEVYAAEOEFQDAEGEFADBEWAEADBFWAABDhBUAwBBhBQAwRFgBAAwRVgAAQ4QVAMCQbcOqqj5SVc9U1YOblr2qqu6uqq8sji/cdNl7q+pEVT1aVdetauAAAOtmmT1WH01y/QuWvSfJPd19VZJ7FudTVVcnOZbktYttPlhVF4yNFgBgjW0bVt39mSTffMHiG5Lcvjh9e5IbNy2/o7uf7e7HkpxIcu3MUAEA1tu5vsbqku5+KkkWxxcvll+W5IlN651cLAMA2PemX7xeWyzrLVesurmqjlfV8VOnTg0PAwBg951rWD1dVZcmyeL4mcXyk0mu2LTe5Ume3OoKuvu27j7a3UcPHz58jsMAAFgf5xpWdyW5aXH6piSf2rT8WFW9rKquTHJVknt3NkQAgPPDoe1WqKqPJXlTkouq6mSS9yf5QJI7q+qdSR5P8rYk6e6HqurOJA8neS7JLd19ekVjBwBYK9uGVXe//QwXvfkM69+a5NadDAoA4Hzkk9cBAIYIKwCAIcIKAGCIsAIAGCKsAACGCCsAgCHCCgBgiLACABgirAAAhggrAIAhwgoAYIiwAgAYIqwAAIYIKwCAIcIKAGCIsAIAGCKsAACGCCsAgCHCCgBgiLACABgirAAAhggrAIAhwgoAYIiwAgAYIqwAAIYIKwCAIcIKAGCIsAIAGCKsAACGCCsAgCHCCgBgiLACABgirAAAhggrAIAhwgoAYIiwAgAYIqwAAIYIKwCAIcIKAGCIsAIAGCKsAACGCCsAgCHCCgBgiLACABgirAAAhggrAIAhwgoAYIiwAgAYIqwAAIYIKwCAIcIKAGCIsAIAGCKsAACGCCsAgCHCCgBgiLACABgirAAAhhzaycZV9bUk305yOslz3X20ql6V5N8kOZLka0n+dnf//s6GCQCw/ib2WP1Id1/T3UcX59+T5J7uvirJPYvzAAD73iqeCrwhye2L07cnuXEFtwEAsHZ2Glad5D9V1X1VdfNi2SXd/VSSLI4v3uFtAACcF3b0Gqskb+zuJ6vq4iR3V9XvLrvhIsRuTpJXv/rVOxwGAMDe29Eeq+5+cnH8TJJfT3Jtkqer6tIkWRw/c4Ztb+vuo9199PDhwzsZBgDAWjjnsKqqP1FVr3j+dJK/leTBJHcluWmx2k1JPrXTQQIAnA928lTgJUl+vaqev55f6+7fqKovJLmzqt6Z5PEkb9v5MAEA1t85h1V3fzXJ67ZY/o0kb97JoAAAzkc+eR0AYIiwAgAYIqwAAIYIKwCAIcIKAGCIsAIAGCKsAACGCCsAgCHCCgBgiLACABgirAAAhggrAIAhwgoAYIiwAgAYIqwAAIYIKwCAIcIKAGCIsAIAGCKsAACGCCsAgCHCCgBgiLACABgirAAAhggrAIAhwgoAYIiwAgAYIqwAAIYIKwCAIcIKAGCIsAIAGCKsAACGCCsAgCHCCgBgiLACABgirAAAhggrAIAhwgoAYIiwAgAYIqwAAIYIKwCAIcIKAGCIsAIAGCKsAACGCCsAgCHCCgBgiLACABgirAAAhggrAIAhwgoAYIiwAgAYIqwAAIYIKwCAIcIKAGDIob0eAKt1442f/a7zn/zkD+/RSAA4SF74709yMP4NElackSgDgLPjqUAAgCH2WJFk/fdO7dYu5XX/PZyvtpq/rfh9A2dr3R63VxZWVXV9kn+e5IIkH+ruD6zqtthbW/1R7+QPfdl/hAFYrXWLlvPBSsKqqi5I8q+S/M0kJ5N8oaru6u6HV3F7B81O997spzvKbvwsOwnHZbZ9se13YidjPF9N/yz7/XezTj+fx7W9N/kf4un73sR17pZV7bG6NsmJ7v5qklTVHUluSCKsVmjdH1j2+56o3fr9r9M878aD6VZ2+oC/6lhedtszWff77rqN70z2ap6Xub6z3X6Z69ytWN6r+/0qbnsVVvXi9cuSPLHp/MnFMgCAfWtVe6xqi2X9XStU3Zzk5sXZP6iqR1c0luddlOTr3z2GFd/iLtrqZ9mNZTsZyybfMzcrup2ztlu/w52MeyfzssR6ZzU3u3Wfmv5b3Kv7ypkssf053We2u43pv6Wdmr5f7NLj2tJzs1ePV9PXt5ePdWf5M+/4frPJnznTBasKq5NJrth0/vIkT25eobtvS3Lbim7/e1TV8e4+ulu3x/LMzfoyN+vJvKwvc7O+dmtuVvVU4BeSXFVVV1bVS5McS3LXim4LAGAtrGSPVXc/V1XvSvIfs/FxCx/p7odWcVsAAOtiZZ9j1d2fTvLpVV3/Odi1px05a+ZmfZmb9WRe1pe5WV+7MjfV3duvBQDAtnxXIADAkH0XVlV1fVU9WlUnquo9W1xeVfUvFpc/UFU/tBfjPIiWmJu/u5iTB6rqt6vqdXsxzoNmu3nZtN5frqrTVfXjuzm+g2yZuamqN1XV/VX1UFX9190e40G1xOPZn6yqf1dVX1rMzU/uxTgPmqr6SFU9U1UPnuHy1TdAd++bQzZeKP8/kvzZJC9N8qUkV79gnbck+Q/Z+KytNyT5/F6P+yAclpybv5rkwsXpHzM36zEvm9b7L9l43eSP7/W4D8JhyfvMK7PxjRavXpy/eK/HfRAOS87N+5L8k8Xpw0m+meSlez32/X5I8teT/FCSB89w+cobYL/tsfr/X6XT3f83yfNfpbPZDUn+dW/4XJJXVtWluz3QA2jbuenu3+7u31+c/Vw2Pv+M1VrmPpMk/yDJx5M8s5uDO+CWmZu/k+QT3f14knS3+dkdy8xNJ3lFVVWSl2cjrJ7b3WEePN39mWz8rs9k5Q2w38Jqma/S8XU7e+Nsf+/vzMb/Klitbeelqi5L8tYkv7yL42K5+8wPJLmwqn6rqu6rqp/YtdEdbMvMzb9M8oPZ+HDsLyd5d3d/Z3eGx4tYeQOs7OMW9si2X6Wz5DrMW/r3XlU/ko2wWr9v19x/lpmXf5bkZ7v7dO2n74Faf8vMzaEkfynJm5N8f5LfqarPdfd/X/XgDrhl5ua6JPcn+dEkr0lyd1X9t+7+1orHxotbeQPst7Da9qt0llyHeUv93qvqLyb5UJIf6+5v7NLYDrJl5uVokjsWUXVRkrdU1XPd/cldGeHBtezj2de7+w+T/GFVfSbJ65IIq9VaZm5+MskHeuOFPSeq6rEkfz7JvbszRM5g5Q2w354KXOardO5K8hOLdwa8Icn/7u6ndnugB9C2c1NVr07yiSTv8D/uXbPtvHT3ld19pLuPJPm3Sf6+qNoVyzyefSrJX6uqQ1X1x5P8lSSP7PI4D6Jl5ubxbOxJTFVdkuTPJfnqro6Sray8AfbVHqs+w1fpVNXfW1z+y9l4V9NbkpxI8n+y8b8KVmzJufmHSf5Ukg8u9o48177MdKWWnBf2wDJz092PVNVvJHkgyXeSfKi7t3ybOXOWvN/84yQfraovZ+Ppp5/t7q/v2aAPiKr6WJI3Jbmoqk4meX+SlyS71wA+eR0AYMh+eyoQAGDPCCsAgCHCCgBgiLACABgirAAAhggrAIAhwgoAYIiwAgAY8v8AGP9c5A9cYSQAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 720x576 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "from pandas.api.types import CategoricalDtype\n",
    "\n",
    "triplets = []\n",
    "for b in bins_of_interest:\n",
    "    bin_i_genes = [i for i in salmon_df.index if f\"{b}_\" in i]\n",
    "    for s in salmon_df.columns:\n",
    "        if not s in ['merger_col', 'gene_name']:\n",
    "            triplets.append((b, s, (salmon_df.loc[bin_i_genes, s] > 0).sum() / len(bin_i_genes)))\n",
    "\n",
    "triplet_frame = pd.DataFrame(triplets)\n",
    "triplet_frame.columns = ['Bin', 'Sample', 'PercentOfGenesExpressed']\n",
    "\n",
    "cat_bin = CategoricalDtype(bins_of_interest)\n",
    "triplet_frame['Bin'] = triplet_frame['Bin'].astype(cat_bin)\n",
    "triplet_frame['PercentOfGenesExpressed'] = triplet_frame['PercentOfGenesExpressed'].astype(float)\n",
    "\n",
    "plt.figure(figsize=(10, 8))\n",
    "\n",
    "n, bins, patches = plt.hist(x=triplet_frame['PercentOfGenesExpressed'], bins=100, color='#0504aa', alpha=0.7, rwidth=0.85)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "reliable-coordination",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['5/17/10_Oxic_3m_1', '5/17/10_Oxic_13m_1', '5/17/10_Oxic_13m_2',\n",
      "       '5/17/10_Oxic_3m_2', '9/21/11_Oxic_17m_2', '6/7/10_Anox_16.5m_1',\n",
      "       '9/21/11_Oxic_17m_1', '6/7/10_Anox_16.5m_2', '7/11/10_Anox_17.5m_1',\n",
      "       '7/11/10_Anox_17.5m_2', '7/11/10_Oxic_3m_1', '7/11/10_Oxic_3m_2',\n",
      "       '8/5/10_Anox_22m_1', '8/5/10_Anox_22m_2', '8/30/10_Oxic_3m_1',\n",
      "       '8/30/10_Oxic_3m_2', '8/30/10_Anox_20m_1', '8/30/10_Anox_20m_2',\n",
      "       '10/18/10_Oxic_13m_1', '10/18/10_Oxic_13m_2', '7/08/11_Oxic_3m_1',\n",
      "       '7/08/11_Oxic_3m_2', '5/24/11_Anox_18m_1', '4/18/11_Oxic_13.5m_1',\n",
      "       '4/18/11_Oxic_13.5m_2', '5/24/11_Anox_18m_2', '6/14/11_Oxic_3m_2',\n",
      "       '6/14/11_Anox_18m_1', '6/14/11_Anox_18m_2', '4_18_11_13.5m_1',\n",
      "       '4_18_11_13.5m_2', '6_14_11_3m_5', '7_8_11_3m_21', '8_30_11_18m_18',\n",
      "       '7_8_11_3m_22', '8_30_11_3m_15', '8_8_11_18m_9_2', '8_30_11_18m_17_2',\n",
      "       '8_8_11_18m_10_2'],\n",
      "      dtype='object')\n",
      "10 29\n",
      "7_8_11_3m_22\t35.74\t5/24/11_Anox_18m_1\n",
      "6_14_11_3m_5\t21.43\t6/14/11_Oxic_3m_2\n",
      "8_8_11_18m_9_2\t57.79\t8/5/10_Anox_22m_1\n",
      "8_30_11_3m_15\t38.94\t9/21/11_Oxic_17m_1\n",
      "7_8_11_3m_21\t56.71\t6/14/11_Oxic_3m_2\n",
      "8_8_11_18m_10_2\t31.01\t8/30/10_Anox_20m_1\n",
      "4_18_11_13.5m_2\t98.29\t7/08/11_Oxic_3m_1\n",
      "8_30_11_18m_17_2\t11.3\t8/30/10_Oxic_3m_1\n",
      "8_30_11_18m_18\tnan\tnan\n",
      "4_18_11_13.5m_1\t96.01\t7/08/11_Oxic_3m_2\n",
      "4/18/11_Oxic_13.5m_1\t98.29\t4_18_11_13.5m_1\n",
      "4/18/11_Oxic_13.5m_2\t95.97\t4_18_11_13.5m_1\n"
     ]
    }
   ],
   "source": [
    "print(salmon_df.columns[:-2])\n",
    "old_cols = list(set(salmon_df.columns[:-2]) & set(hewson_translator.values()))\n",
    "new_cols = list(set(salmon_df.columns[:-2]) - set(hewson_translator.values()))\n",
    "print(len(new_cols), len(old_cols))\n",
    "sample_corrs = salmon_df[new_cols + old_cols].corr()*100\n",
    "for samp_c in new_cols:\n",
    "    print(f\"{samp_c}\\t{round(sample_corrs.loc[samp_c, old_cols].max(), 2)}\\t{sample_corrs.loc[samp_c, old_cols].idxmax()}\")\n",
    "\n",
    "\n",
    "print(f\"{hewson_translator['SRR988008']}\\t{round(sample_corrs.loc[hewson_translator['SRR988008'], new_cols].max(), 2)}\\t{sample_corrs.loc[samp_c, new_cols].idxmax()}\")\n",
    "print(f\"{hewson_translator['SRR988009']}\\t{round(sample_corrs.loc[hewson_translator['SRR988009'], new_cols].max(), 2)}\\t{sample_corrs.loc[samp_c, new_cols].idxmax()}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "suitable-karen",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(201313, 36) (1051, 34)\n",
      "10/18/10_Oxic_13m_1\n",
      "10/18/10_Oxic_13m_2\n",
      "4/18/11_Oxic_13.5m_1\n",
      "4/18/11_Oxic_13.5m_2\n",
      "5/17/10_Oxic_13m_1\n",
      "5/17/10_Oxic_13m_2\n",
      "5/17/10_Oxic_3m_1\n",
      "5/17/10_Oxic_3m_2\n",
      "5/24/11_Anox_18m_1\n",
      "5/24/11_Anox_18m_2\n",
      "6/14/11_Anox_18m_1\n",
      "6/14/11_Anox_18m_2\n",
      "6/14/11_Oxic_3m_1\n",
      "6/14/11_Oxic_3m_2\n",
      "6/7/10_Anox_16.5m_1\n",
      "6/7/10_Anox_16.5m_2\n",
      "7/11/10_Anox_17.5m_1\n",
      "7/11/10_Anox_17.5m_2\n",
      "7/11/10_Oxic_3m_1\n",
      "7/11/10_Oxic_3m_2\n",
      "7/8/11_Oxic_3m_1\n",
      "7/8/11_Oxic_3m_2\n",
      "8/30/10_Anox_20m_1\n",
      "8/30/10_Anox_20m_2\n",
      "8/30/10_Oxic_3m_1\n",
      "8/30/10_Oxic_3m_2\n",
      "8/30/11_Oxic_18m\n",
      "8/30/11_Oxic_3m\n",
      "8/5/10_Anox_22m_1\n",
      "8/5/10_Anox_22m_2\n",
      "8/8/11_Anox_18m_1\n",
      "8/8/11_Anox_18m_2\n",
      "9/21/11_Oxic_17m_1\n",
      "9/21/11_Oxic_17m_2\n",
      "gene_name\n",
      "merger_col\n"
     ]
    }
   ],
   "source": [
    "new_col_renamer = {'4_18_11_13.5m_1': '4/18/11_Oxic_13.5m_1',\n",
    "                   '4_18_11_13.5m_2': '4/18/11_Oxic_13.5m_2',\n",
    "                   '6_14_11_3m_5': '6/14/11_Oxic_3m_1',\n",
    "                   '7_8_11_3m_21': '7/8/11_Oxic_3m_1',\n",
    "                   '7_8_11_3m_22': '7/8/11_Oxic_3m_2',\n",
    "                   '8_30_11_3m_15': '8/30/11_Oxic_3m',\n",
    "                   '8_8_11_18m_9_2': '8/8/11_Anox_18m_1',\n",
    "                   '8_30_11_18m_17_2': '8/30/11_Oxic_18m',\n",
    "                   '8_8_11_18m_10_2': '8/8/11_Anox_18m_2'}\n",
    "\n",
    "mislabelled_or_crap = ['7/08/11_Oxic_3m_2', \"7/08/11_Oxic_3m_1\", \n",
    "                       '4/18/11_Oxic_13.5m_1', '4/18/11_Oxic_13.5m_2', \n",
    "                       \"8_30_11_18m_18\"]\n",
    "\n",
    "salmon_df_ = salmon_df.drop(mislabelled_or_crap, axis=1).rename(columns=new_col_renamer)\n",
    "salmon_dfs_ = {i:j.drop(mislabelled_or_crap, axis=1).rename(columns=new_col_renamer) for i, j in salmon_dfs.items()}\n",
    "\n",
    "print(salmon_df_.shape, (list(salmon_dfs_.values())[0]).shape)\n",
    "print(\"\\n\".join(sorted(list(salmon_df_.columns))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "clean-bookmark",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CB33.bin.167\t1\n",
      "CBrest.bin.390\t1\n",
      "CBrest.bin.385\t1\n",
      "CBrest.bin.111\t2\n",
      "CB33.bin.201\t2\n",
      "CB33.bin.212\t2\n",
      "CB33.bin.32\t2\n",
      "CB33.bin.34\t2\n",
      "CBrest.bin.370\t2\n",
      "CBrest.bin.331\t2\n",
      "CB33.bin.146\t3\n",
      "CB33.bin.222\t3\n",
      "CB33.bin.56\t3\n",
      "CBrest.bin.189\t3\n",
      "CBrest.bin.196\t3\n",
      "CB33.bin.2\t3\n",
      "CBrest.bin.6\t3\n",
      "CB33.bin.175\t3\n",
      "CB33.bin.82\t3\n",
      "CB33.bin.86\t3\n",
      "CB33.bin.90\t3\n",
      "CB33.bin.186\t3\n",
      "CB33.bin.95\t4\n",
      "CBrest.bin.25\t4\n",
      "CBrest.bin.322\t4\n",
      "CBrest.bin.344\t4\n",
      "CB33.bin.179\t4\n",
      "CB33.bin.196\t4\n",
      "CB33.bin.202\t4\n",
      "CB33.bin.207\t4\n",
      "CB33.bin.218\t4\n",
      "CBrest.bin.327\t4\n",
      "CBrest.bin.223\t4\n",
      "CBrest.bin.374\t4\n",
      "CB33.bin.118\t5\n",
      "CB33.bin.219\t5\n",
      "CB33.bin.89\t5\n",
      "CBrest.bin.26\t5\n",
      "CBrest.bin.63\t5\n",
      "CBrest.bin.94\t5\n",
      "CBrest.bin.377\t5\n",
      "CB33.bin.84\t5\n",
      "CB33.bin.216\t5\n",
      "CB33.bin.227\t5\n",
      "CB33.bin.229\t5\n",
      "CBrest.bin.246\t5\n",
      "CB33.bin.61\t5\n",
      "CBrest.bin.394\t5\n",
      "CB33.bin.133\t6\n",
      "CB33.bin.174\t6\n",
      "CB33.bin.235\t6\n",
      "CB33.bin.191\t6\n",
      "CBrest.bin.118\t6\n",
      "CB33.bin.85\t6\n",
      "CBrest.bin.355\t6\n",
      "CBrest.bin.358\t6\n",
      "CBrest.bin.61\t6\n",
      "CBrest.bin.84\t6\n",
      "CBrest.bin.96\t6\n",
      "CBrest.bin.388\t6\n",
      "CBrest.bin.40\t6\n",
      "CB33.bin.182\t6\n",
      "CB33.bin.194\t6\n",
      "CB33.bin.233\t6\n",
      "CBrest.bin.263\t6\n",
      "CBrest.bin.368\t6\n",
      "CB33.bin.169\t7\n",
      "CB33.bin.203\t7\n",
      "CB33.bin.236\t7\n",
      "CBrest.bin.113\t7\n",
      "CBrest.bin.395\t7\n",
      "CB33.bin.63\t7\n",
      "CBrest.bin.333\t7\n",
      "CB33.bin.149\t8\n",
      "CB33.bin.168\t8\n",
      "CB33.bin.172\t8\n",
      "CBrest.bin.174\t8\n",
      "CBrest.bin.191\t8\n",
      "CBrest.bin.346\t8\n",
      "CBrest.bin.359\t8\n",
      "CBrest.bin.408\t8\n",
      "CB33.bin.104\t9\n",
      "CBrest.bin.252\t9\n",
      "CBrest.bin.389\t9\n",
      "CB33.bin.142\t10\n",
      "CB33.bin.185\t10\n",
      "CBrest.bin.188\t10\n",
      "CBrest.bin.349\t10\n",
      "CBrest.bin.405\t10\n",
      "CBrest.bin.375\t10\n",
      "CBrest.bin.384\t10\n",
      "CB33.bin.111\t12\n"
     ]
    }
   ],
   "source": [
    "sample_cnt = []\n",
    "for k in avail_bins:\n",
    "    nocrap = [i for i in avail_bins[k] if not i in mislabelled_or_crap]\n",
    "    renamed = [new_col_renamer[i] if i in new_col_renamer else i for i in nocrap]\n",
    "    derepd = [i[:-2] if i[-2:] == '_1' or i[-2:] == '_2' else i for i in renamed ]\n",
    "    sample_cnt.append((k, len(set(derepd))))\n",
    "\n",
    "    \n",
    "print(\"\\n\".join([f\"{i[0]}\\t{i[1]}\" for i in sorted(sample_cnt, key=lambda x: x[1])]))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "naughty-young",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10/18/10_Oxic_13m_1 22\n",
      "10/18/10_Oxic_13m_2 21\n",
      "4/18/11_Oxic_13.5m_1 13\n",
      "4/18/11_Oxic_13.5m_2 1\n",
      "5/17/10_Oxic_13m_1 14\n",
      "5/17/10_Oxic_13m_2 15\n",
      "5/17/10_Oxic_3m_1 20\n",
      "5/17/10_Oxic_3m_2 20\n",
      "5/24/11_Anox_18m_1 2\n",
      "5/24/11_Anox_18m_2 1\n",
      "6/14/11_Anox_18m_1 2\n",
      "6/14/11_Anox_18m_2 11\n",
      "6/14/11_Oxic_3m_2 7\n",
      "6/7/10_Anox_16.5m_1 48\n",
      "6/7/10_Anox_16.5m_2 47\n",
      "6/14/11_Oxic_3m_1 1\n",
      "7/11/10_Anox_17.5m_1 64\n",
      "7/11/10_Anox_17.5m_2 64\n",
      "7/11/10_Oxic_3m_1 68\n",
      "7/11/10_Oxic_3m_2 66\n",
      "7/8/11_Oxic_3m_1 6\n",
      "7/8/11_Oxic_3m_2 4\n",
      "8/30/10_Anox_20m_1 59\n",
      "8/30/10_Anox_20m_2 58\n",
      "8/30/10_Oxic_3m_1 50\n",
      "8/30/10_Oxic_3m_2 50\n",
      "8/5/10_Anox_22m_1 41\n",
      "8/5/10_Anox_22m_2 41\n",
      "8/30/11_Oxic_18m 2\n",
      "8/30/11_Oxic_3m 9\n",
      "8/8/11_Anox_18m_2 13\n",
      "8/8/11_Anox_18m_1 34\n",
      "9/21/11_Oxic_17m_1 23\n",
      "9/21/11_Oxic_17m_2 9\n"
     ]
    }
   ],
   "source": [
    "for s in sorted(avail_samples):\n",
    "    if not s in mislabelled_or_crap:\n",
    "        b = avail_samples[s]\n",
    "        if s in new_col_renamer:\n",
    "            s = new_col_renamer[s]\n",
    "        print(s, len(b))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "british-comedy",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19\n",
      "5/17/10_Oxic_3m: {'5/17/10_Oxic_3m_1', '5/17/10_Oxic_3m_2'}\n",
      "5/17/10_Oxic_13m: {'5/17/10_Oxic_13m_1', '5/17/10_Oxic_13m_2'}\n",
      "9/21/11_Oxic_17m: {'9/21/11_Oxic_17m_1', '9/21/11_Oxic_17m_2'}\n",
      "6/7/10_Anox_16.5m: {'6/7/10_Anox_16.5m_1', '6/7/10_Anox_16.5m_2'}\n",
      "7/11/10_Anox_17.5m: {'7/11/10_Anox_17.5m_1', '7/11/10_Anox_17.5m_2'}\n",
      "7/11/10_Oxic_3m: {'7/11/10_Oxic_3m_1', '7/11/10_Oxic_3m_2'}\n",
      "8/5/10_Anox_22m: {'8/5/10_Anox_22m_2', '8/5/10_Anox_22m_1'}\n",
      "8/30/10_Oxic_3m: {'8/30/10_Oxic_3m_2', '8/30/10_Oxic_3m_1'}\n",
      "8/30/10_Anox_20m: {'8/30/10_Anox_20m_2', '8/30/10_Anox_20m_1'}\n",
      "10/18/10_Oxic_13m: {'10/18/10_Oxic_13m_1', '10/18/10_Oxic_13m_2'}\n",
      "5/24/11_Anox_18m: {'5/24/11_Anox_18m_1', '5/24/11_Anox_18m_2'}\n",
      "6/14/11_Oxic_3m: {'6/14/11_Oxic_3m_1', '6/14/11_Oxic_3m_2'}\n",
      "6/14/11_Anox_18m: {'6/14/11_Anox_18m_2', '6/14/11_Anox_18m_1'}\n",
      "4/18/11_Oxic_13.5m: {'4/18/11_Oxic_13.5m_2', '4/18/11_Oxic_13.5m_1'}\n",
      "7/8/11_Oxic_3m: {'7/8/11_Oxic_3m_1', '7/8/11_Oxic_3m_2'}\n",
      "8/30/11_Oxic_3m: {'8/30/11_Oxic_3m'}\n",
      "8/8/11_Anox_18m: {'8/8/11_Anox_18m_2', '8/8/11_Anox_18m_1'}\n",
      "8/30/11_Oxic_18m: {'8/30/11_Oxic_18m'}\n",
      "extras: {'gene_name', 'merger_col'}\n"
     ]
    }
   ],
   "source": [
    "dereplication_groups = {}\n",
    "for c in salmon_df_.columns:\n",
    "    if c.endswith(\"_2\") or c.endswith(\"_1\"):\n",
    "        dereplication_groups[c[:-2]] = {i for i in salmon_df_.columns if i.startswith(c[:-2])}\n",
    "    elif 'ox' in c.lower():\n",
    "        dereplication_groups[c] = {c}\n",
    "    else:\n",
    "        dereplication_groups.setdefault('extras', set()).add(c)\n",
    "\n",
    "print(len(dereplication_groups))\n",
    "print(\"\\n\".join([f\"{k}: {v}\" for k, v in dereplication_groups.items()]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "convinced-assembly",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_corrs = salmon_df_.corr()*100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "adequate-knight",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(201313, 20)\n",
      "Index(['5/17/10_Oxic_3m', '5/17/10_Oxic_13m', '9/21/11_Oxic_17m',\n",
      "       '6/7/10_Anox_16.5m', '7/11/10_Anox_17.5m', '7/11/10_Oxic_3m',\n",
      "       '8/5/10_Anox_22m', '8/30/10_Oxic_3m', '8/30/10_Anox_20m',\n",
      "       '10/18/10_Oxic_13m', '5/24/11_Anox_18m', '6/14/11_Oxic_3m',\n",
      "       '6/14/11_Anox_18m', '4/18/11_Oxic_13.5m', '7/8/11_Oxic_3m',\n",
      "       '8/30/11_Oxic_3m', '8/8/11_Anox_18m', '8/30/11_Oxic_18m', 'gene_name',\n",
      "       'merger_col'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "dereplicated_srs = []\n",
    "for new_name, old_cols in dereplication_groups.items():\n",
    "    if len(old_cols) == 1 or new_name == 'extras':\n",
    "        dereplicated_srs.append(salmon_df_[old_cols].copy())\n",
    "    else:\n",
    "        new_srs = salmon_df_[old_cols].sum(1) / 2\n",
    "        new_srs.name = new_name\n",
    "        dereplicated_srs.append(new_srs)\n",
    "            \n",
    "derep_df = pd.concat(dereplicated_srs, axis=1, verify_integrity=True)      \n",
    "print(derep_df.shape)\n",
    "print(derep_df.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "adaptive-pakistan",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['5/17/10_Oxic_3m', '5/17/10_Oxic_13m', '9/21/11_Oxic_17m', '6/7/10_Anox_16.5m', '7/11/10_Anox_17.5m', '7/11/10_Oxic_3m', '8/5/10_Anox_22m', '8/30/10_Oxic_3m', '8/30/10_Anox_20m', '10/18/10_Oxic_13m', '5/24/11_Anox_18m', '6/14/11_Oxic_3m', '6/14/11_Anox_18m', '4/18/11_Oxic_13.5m', '7/8/11_Oxic_3m', '8/30/11_Oxic_3m', '8/8/11_Anox_18m', '8/30/11_Oxic_18m'] 18\n",
      "201313\n",
      "201313\n"
     ]
    }
   ],
   "source": [
    "sample_cols = list(derep_df.columns[:-3])\n",
    "print(sample_cols, len(sample_cols))\n",
    "annot_and_quant = derep_df.copy()\n",
    "annot_and_quant['merger_col'] = annot_and_quant['merger_col'].apply(lambda x: (x[0], x[1] + \"_\" + x[2], x[3], x[4]))\n",
    "print(len(set(annot_and_quant['merger_col']) & set(pathway_df['merger_col'])))\n",
    "print(len(set(annot_and_quant['merger_col'])))\n",
    "annot_and_quant = annot_and_quant.merge(pathway_df, on='merger_col', validate='1:m').copy()\n",
    "descript_cols = ['genome', 'gene', 'locus_tag', 'product', 'query_id']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "compliant-score",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7944\n"
     ]
    }
   ],
   "source": [
    "reverse_gene_sets = {vv:k for k, v in gene_sets.items() for vv in v}\n",
    "annot_and_quant['annotation_code'] = annot_and_quant.identifier.map(reverse_gene_sets)\n",
    "print(annot_and_quant['annotation_code'].notnull().sum())\n",
    "annot_and_quant = annot_and_quant[annot_and_quant.annotation_code.notnull()].copy()\n",
    "#annot_and_quant.sample(1).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "million-cardiff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CB33.bin.175\t('nirK/nirS+norBC', 'dsrAB')\t['8/8/11_Anox_18m', '7/11/10_Anox_17.5m']\n",
      "CB33.bin.202\t('nirK/nirS+norBC', 'dsrAB')\t['8/30/10_Anox_20m', '8/8/11_Anox_18m', '7/11/10_Anox_17.5m']\n",
      "CB33.bin.236\t('nirK/nirS+norBC', 'dsrAB')\t['8/30/10_Anox_20m', '6/7/10_Anox_16.5m', '6/14/11_Anox_18m', '8/8/11_Anox_18m', '7/11/10_Oxic_3m', '7/11/10_Anox_17.5m']\n",
      "CB33.bin.222\t('nirK/nirS+norBC', 'dsrAB')\t['7/11/10_Anox_17.5m']\n",
      "CB33.bin.216\t('nirK/nirS+norBC', 'dsrAB')\t['8/30/10_Anox_20m', '7/11/10_Anox_17.5m', '8/8/11_Anox_18m', '8/5/10_Anox_22m']\n",
      "CB33.bin.196\t('nirK/nirS+norBC', 'dsrAB')\t['8/30/10_Anox_20m', '8/8/11_Anox_18m', '8/5/10_Anox_22m']\n",
      "CB33.bin.175\t('napAB+narIJ', 'dsrAB')\t['7/11/10_Anox_17.5m']\n",
      "CB33.bin.202\t('napAB+narIJ', 'dsrAB')\t['8/30/10_Anox_20m', '8/8/11_Anox_18m']\n",
      "CB33.bin.236\t('napAB+narIJ', 'dsrAB')\t['8/30/10_Anox_20m', '6/14/11_Anox_18m', '8/8/11_Anox_18m', '7/11/10_Anox_17.5m']\n",
      "CB33.bin.216\t('napAB+narIJ', 'dsrAB')\t['8/30/10_Anox_20m', '8/5/10_Anox_22m']\n",
      "CBrest.bin.358\t('napAB+narIJ', 'dsrAB')\t['6/14/11_Anox_18m', '8/30/10_Anox_20m', '6/7/10_Anox_16.5m']\n",
      "CB33.bin.196\t('napAB+narIJ', 'dsrAB')\t['8/30/10_Anox_20m', '8/8/11_Anox_18m', '8/5/10_Anox_22m']\n",
      "CB33.bin.236\t('nirABD', 'dsrAB')\t['8/30/10_Anox_20m', '6/7/10_Anox_16.5m', '7/11/10_Oxic_3m', '7/11/10_Anox_17.5m']\n",
      "CB33.bin.222\t('nirABD', 'dsrAB')\t['7/11/10_Anox_17.5m']\n",
      "CB33.bin.216\t('nirABD', 'dsrAB')\t['8/30/10_Anox_20m', '7/11/10_Anox_17.5m', '8/8/11_Anox_18m', '8/5/10_Anox_22m']\n",
      "CB33.bin.196\t('nirABD', 'dsrAB')\t['8/30/10_Anox_20m', '8/8/11_Anox_18m']\n"
     ]
    }
   ],
   "source": [
    "hq_bins = set(qualtax_df.index[(qualtax_df.Completeness > 80) & (qualtax_df.Contamination < 20)])\n",
    "key2_genes = gene_sets['dsrAB']\n",
    "key3_genes = gene_sets['rbcSL']\n",
    "\n",
    "key_indexes = list()\n",
    "for nit_gene in ['nirK/nirS+norBC', 'nrfAH', 'napAB+narIJ', 'nirABD']:\n",
    "    key_pair = (nit_gene, 'dsrAB')\n",
    "    these_genomes = annot_mat.index[annot_mat[list(key_pair)].sum(1) == 2]\n",
    "    these_genomes = set(these_genomes) & hq_bins\n",
    "    if len(these_genomes):\n",
    "        for this_genome in these_genomes:\n",
    "            key1_genes = gene_sets[key_pair[0]]\n",
    "            key1_sub_genes = (annot_and_quant.identifier.isin(list(key1_genes)) &\n",
    "                             (annot_and_quant.genome == this_genome))\n",
    "            key_idx = list(annot_and_quant.index[key1_sub_genes])\n",
    "            key1_nz_expression = annot_and_quant.loc[key1_sub_genes, sample_cols].sum() > 0\n",
    "\n",
    "            key2_sub_genes = (annot_and_quant.identifier.isin(list(key2_genes)) & \n",
    "                             (annot_and_quant.genome == this_genome))\n",
    "            key_idx += list(annot_and_quant.index[key2_sub_genes])\n",
    "            key2_nz_expression = annot_and_quant.loc[key2_sub_genes, sample_cols].sum() > 0\n",
    "            \n",
    "            key3_sub_genes = (annot_and_quant.identifier.isin(list(key3_genes)) & \n",
    "                             (annot_and_quant.genome == this_genome))\n",
    "            key_idx += list(annot_and_quant.index[key3_sub_genes])\n",
    "            key3_nz_expression = annot_and_quant.loc[key3_sub_genes, sample_cols].sum() > 0\n",
    "            \n",
    "            stacked_bools = pd.concat([key1_nz_expression, key2_nz_expression], axis=1).T\n",
    "            stacked_bools2 = pd.concat([key1_nz_expression, key2_nz_expression, key3_nz_expression], axis=1).T\n",
    "            \n",
    "            simultaneous_cols = set(stacked_bools.columns[stacked_bools.sum() == 2])\n",
    "            simultaneous_cols.update(set(stacked_bools.columns[stacked_bools2.sum() == 3]))\n",
    "            \n",
    "            if len(simultaneous_cols):\n",
    "                key_indexes += key_idx\n",
    "                print(\"\\t\".join([this_genome, str(key_pair), str(list(simultaneous_cols))]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "cardiac-programmer",
   "metadata": {},
   "outputs": [],
   "source": [
    "excluded_cols = ['merger_col', 'merger_col_2', 'metacyc_target', 'ClusterID', 'GeneID']\n",
    "included_cols = [i for i in annot_and_quant.columns if not i in excluded_cols and not i in sample_cols] + sample_cols\n",
    "annot_and_quant[included_cols].sort_values(by=['genome', 'annotation_code']).to_csv('~/OneDrive - Johns Hopkins/CB_16S_Shotgun_Manuscript/Data/Functional_Gene_Expression_and_annotations_all.txt', sep=\"\\t\")\n",
    "annot_and_quant.loc[set(key_indexes), included_cols].sort_values(by=['genome', 'annotation_code']).to_csv('~/OneDrive - Johns Hopkins/CB_16S_Shotgun_Manuscript/Data/Functional_Gene_Expression_and_annotations_no3r_rdsr_subset.txt', sep=\"\\t\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "voluntary-joyce",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "configured-cornwall",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_report(qualtax_df, complete_list, cbTr_abnd2, cb33_abnd2, salmon_df, copy_num_df, gene_id_lookup3, gene_sets, gene_id_lookup2):\n",
    "    \n",
    "    report_df = qualtax_df.loc[complete_list, ['Completeness', 'Contamination', 'N50 (scaffolds)']]\n",
    "    \n",
    "    addl_columns = ['taxonomic leaf rank', 'taxonomic leaf name', 'max copy number', \n",
    "                    'genes mapped', \n",
    "                    'genes with sum >5x rpm expression', 'genes with mean >5x rpm expression', \n",
    "                    'median copy number (target genes)', 'sum expression (target genes)',\n",
    "                    '# of enzymes max >1x (exp)', 'enzymes max >1x (exp)',\n",
    "                    '# of enzymes max >10x (exp)', 'enzymes max >10x (exp)',\n",
    "                    '# of enzymes max >1 rpm (copy)', 'enzymes max >1 rpm (copy)',\n",
    "                    '# of enzymes max >10 rpm (copy)', 'enzymes max >10 rpm (copy)']\n",
    "    \n",
    "    for c in addl_columns:\n",
    "        if c.startswith(\"enzymes\"):\n",
    "            report_df[c] = pd.Series(index=report_df.index, dtype=str, data=['']*len(report_df))\n",
    "        else:\n",
    "            report_df[c] = pd.Series(index=report_df.index, dtype=str)\n",
    "    \n",
    "    cbTrexp = [i for i in cbTr_abnd2.columns if not 'control' in i.lower()]\n",
    "    cb33exp = [i for i in cb33_abnd2.columns if not 'ctrl' in i.lower()]\n",
    "    \n",
    "    numeric_exp = [i for i in salmon_df.columns if not i in ['merger_col', 'gene_name']]\n",
    "    \n",
    "    exp_keys = set(salmon_df.index)\n",
    "    copy_num_keys = set(copy_num_df.index)\n",
    "    exp_keys_translated = set([i for i in exp_keys if i in gene_id_lookup3 and gene_id_lookup3[i] in copy_num_keys])\n",
    "    print(len(exp_keys_translated), 'expression keys available')\n",
    "    \n",
    "    for bof in complete_list:\n",
    "        print(\"getting lowest classification\")\n",
    "        done_flag = False\n",
    "        for c in ['Species', 'Genus', 'Family', 'Order', 'Class', 'Phylum', 'Domain']:\n",
    "            cname = str(qualtax_df.loc[bof, c])\n",
    "            if (cname.lower() in ['nan', '']):\n",
    "                pass            \n",
    "            elif not done_flag:\n",
    "                done_flag = True\n",
    "                report_df.loc[bof, 'taxonomic leaf rank'] = c\n",
    "                report_df.loc[bof, 'taxonomic leaf name'] = cname\n",
    "        \n",
    "        if bof in (salmon_dfs.keys()):\n",
    "            pass\n",
    "            print(bof, report_df.loc[bof, 'taxonomic leaf name'])\n",
    "        else:\n",
    "            continue        \n",
    "        \n",
    "        if bof.startswith('CBrest'):\n",
    "            slice_data = cbTr_abnd2.loc[bof, cbTrexp].copy()\n",
    "        elif bof.startswith('CB33'):\n",
    "            slice_data = cb33_abnd2.loc[bof, cb33exp].copy()\n",
    "            \n",
    "        report_df.loc[bof, 'max copy number'] = np.max(slice_data)\n",
    "        report_df.loc[bof, 'num samples w/copy >1x'] = np.sum((slice_data > 1).astype(int))\n",
    "        print(report_df.loc[bof, ['max copy number', 'num samples w/copy >1x']])\n",
    "        all_exp_found = exp_keys.intersection(gene_id_lookup2[bof])\n",
    "        #print('Genes found in copy num + expression data', len(all_exp_found))\n",
    "        \n",
    "        report_df.loc[bof, 'genes with sum >5x rpm expression'] = (salmon_df.loc[all_exp_found, numeric_exp].sum(1) > 5).sum()\n",
    "        report_df.loc[bof, 'genes with mean >5x rpm expression'] = (salmon_df.loc[all_exp_found, numeric_exp].mean(1) > 5).sum()\n",
    "        report_df.loc[bof, 'genes mapped'] = len(all_exp_found)\n",
    "        report_df.loc[bof, 'expression variance'] = salmon_df.loc[all_exp_found, numeric_exp].sum(1).var() / 1e9\n",
    "        report_df.loc[bof, '# of genes expressed'] = (salmon_df.loc[all_exp_found, numeric_exp] > 0).sum().sum()\n",
    "        \n",
    "        #print(report_df.loc[bof, ['genes with sum >5x rpm expression', 'genes with mean >5x rpm expression', 'genes mapped', 'expression variance']])\n",
    "         \n",
    "        copy_list = []\n",
    "        expression_list = []\n",
    "        for k, pre_keys in gene_sets.items():\n",
    "            local_keys = [i for i in pre_keys if i.startswith(bof+\"_\")]\n",
    "            expindx = set(local_keys) & set(salmon_df.index)\n",
    "            if len(expindx):\n",
    "                expvalue = np.max(salmon_df.loc[expindx, numeric_exp].sum())\n",
    "                expression_list.append(expvalue)\n",
    "                if expvalue >= 1:\n",
    "                    report_df.loc[bof, 'enzymes max >1x (exp)'] = \" \".join([report_df.loc[bof, 'enzymes max >1x (exp)'], f\"{k}\"])\n",
    "                if expvalue >= 10:\n",
    "                    report_df.loc[bof, 'enzymes max >10x (exp)'] = \" \".join([report_df.loc[bof, 'enzymes max >10x (exp)'], f\"{k}\"])\n",
    "            \n",
    "            cnindx = set([gene_id_lookup3[lk] for lk in set(local_keys) & exp_keys_translated]) \n",
    "            if len(cnindx):\n",
    "                cnvalue = np.max(copy_num_df.loc[cnindx, :].sum())\n",
    "                copy_list.append(cnvalue)\n",
    "                if cnvalue >= 1:\n",
    "                    report_df.loc[bof, 'enzymes max >1 rpm (copy)'] = \" \".join([report_df.loc[bof, 'enzymes max >1 rpm (copy)'], f\"{k}\"])\n",
    "                if cnvalue >= 10:\n",
    "                    report_df.loc[bof, 'enzymes max >10 rpm (copy)'] = \" \".join([report_df.loc[bof, 'enzymes max >10 rpm (copy)'], f\"{k}\"])\n",
    "        \n",
    "        #print(report_df.loc[bof, ['enzymes max >1x (exp)', 'enzymes max >10x (exp)', \n",
    "        #                         'enzymes max >1 rpm (copy)', 'enzymes max >10 rpm (copy)']].T)\n",
    "        \n",
    "        if len(copy_list):\n",
    "            report_df.loc[bof, 'median copy number (target genes)'] = np.median(np.array(copy_list))\n",
    "        \n",
    "        if len(expression_list):\n",
    "            report_df.loc[bof, 'sum expression (target genes)'] = np.sum(np.array(expression_list))\n",
    "\n",
    "        #print(report_df.loc[bof, ['median copy number (target genes)', 'sum expression (target genes)']].T)\n",
    "        \n",
    "        report_df.loc[bof, '# of enzymes max >1x (exp)'] = len(report_df.loc[bof, 'enzymes max >1x (exp)'].split())\n",
    "        report_df.loc[bof, '# of enzymes max >10x (exp)'] = len(report_df.loc[bof, 'enzymes max >10x (exp)'].split())\n",
    "        report_df.loc[bof, '# of enzymes max >1 rpm (copy)'] = len(report_df.loc[bof, 'enzymes max >1 rpm (copy)'].split())\n",
    "        report_df.loc[bof, '# of enzymes max >10 rpm (copy)'] = len(report_df.loc[bof, 'enzymes max >10 rpm (copy)'].split())\n",
    "        \n",
    "    cols_to_rank = ['max copy number', 'num samples w/copy >1x', 'sum expression (target genes)',\n",
    "                    '# of genes expressed']\n",
    "    \n",
    "    rank_cols = [i+\" Rank\" for i in cols_to_rank]\n",
    "    \n",
    "    for cr, rc in zip(cols_to_rank, rank_cols):\n",
    "        report_df[rc] = report_df[cr].rank()\n",
    "    \n",
    "    report_df['mean rank'] = report_df[rank_cols].mean(1)\n",
    "    return report_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "mounted-power",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "copy_data_file = \"/Volumes/KeithSSD/SulFox/data/gene_expression/gene_copy_number_data.txt\"\n",
    "copy_duplicates_file = \"/Volumes/KeithSSD/SulFox/data/gene_expression/copy_number_duplicate_clusters.txt\"\n",
    "copy_num_df = load_cn_data(copy_data_file, copy_duplicates_file)\n",
    "\n",
    "gene_id_lookup2 = pathway_df[['genome', 'merger_col']].dropna().groupby('genome').agg(set)['merger_col'].to_dict()\n",
    "gene_id_lookup2 = {i:{\"_\".join(kk) for kk in k} for i, k in gene_id_lookup2.items()}\n",
    "gene_id_lookup3 = {i:j for i, j in pathway_df[['identifier', 'GeneID']].dropna().values}\n",
    "report_df = make_report(qualtax_df, complete_list, cbTr_abnd2, cb33_abnd2, derep_df, \n",
    "                        copy_num_df, gene_id_lookup3, gene_sets, gene_id_lookup2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "declared-stroke",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_cats = pd.read_csv('/Volumes/KeithSSD/SulFox/data/genes_by_pathway.txt', header=None, sep=\"\\t\")\n",
    "\n",
    "n_genes = np.array(list(gene_cats.iloc[15:23, 1]))\n",
    "print(n_genes)\n",
    "s_genes = np.array(list(gene_cats.iloc[23:-7, 1]))\n",
    "print(s_genes)\n",
    "add_genes_N = lambda x: \" \".join(list(n_genes[annot_mat.loc[x, n_genes]]))\n",
    "add_genes_S = lambda x: \" \".join(list(s_genes[annot_mat.loc[x, s_genes]]))\n",
    "report_df2 = report_df.sort_values(by='mean rank', ascending=False)\n",
    "srs1df = pd.Series({i:add_genes_N(i) for i in report_df.index}, name='N')\n",
    "srs1df.name = 'N genes'\n",
    "srs2df = pd.Series({i:add_genes_S(i) for i in report_df.index}, name='S')\n",
    "srs1df.name = 'S genes'\n",
    "report_df3 = report_df2.copy().join(srs1df).join(srs2df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "mighty-boundary",
   "metadata": {},
   "outputs": [],
   "source": [
    "outdir = '/Users/login/OneDrive - Johns Hopkins/CB_16S_Shotgun_Manuscript/Supplemental Materials'\n",
    "writer = pd.ExcelWriter(f\"{outdir}/bin_prioritization_report.xlsx\", engine='xlsxwriter')\n",
    "report_df3.to_excel(writer)\n",
    "writer.save()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "vital-equation",
   "metadata": {},
   "outputs": [],
   "source": [
    "qualtax_df.loc[qualtax_df['Genus'] == 'AAA536-G10'].T\n",
    "\n",
    "cbTr_abnd2.loc['CBrest.bin.174', :] > 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "missing-british",
   "metadata": {},
   "outputs": [],
   "source": [
    "bins_of_focus = [\"CBrest.bin.375\", # Thiohalospirales\n",
    "                 \"CBrest.bin.263\", # Thioglobaceae\n",
    "                 \"CBrest.bin.246\", # Nitrosopumilus\n",
    "                 \"CB33.bin.142\", # AAA164-E04\n",
    "                 \"CB33.bin.111\", # Vulcanococcus sp000179255\n",
    "                 \"CB33.bin.196\", # GCA-001735895\n",
    "                 \"CB33.bin.182\", # UBA8477\n",
    "                 \"CBrest.bin.358\", # Marivivens\n",
    "                 \"CB33.bin.236\", # Sedimenticola \n",
    "                 'CBrest.bin.385'] # HIMB11 sp003486095\n",
    "\n",
    "bins_of_focus = set(bins_of_focus) | set([\"CB33.bin.222\", # Magnetovibrionaceae\n",
    "                                          \"CB33.bin.175\", # UBA3470\n",
    "                                          \"CB33.bin.202\", # GCF-002020805\n",
    "                                          \"CBrest.bin.40\"]) # Sedimenticolaceae\n",
    "\n",
    "qualtax_df.loc[bins_of_focus, ['Genus', 'Family', 'Order', 'Class', 'Phylum']]\n",
    "(annot_mat.loc[bins_of_focus, :].T.sum(1) > 0).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "future-nickname",
   "metadata": {},
   "outputs": [],
   "source": [
    "special_gene_sets = set([annot for bof in bins_of_focus for annot in annot_mat.columns[annot_mat.loc[bof, :]]])\n",
    "special_gene_sets = special_gene_sets - set(['', 'bcl_synth', 'taurine_degradation'])\n",
    "\n",
    "housekeepers = [\"proC\", \"recA\", \"rpoD\", \"rho\", \"glyA\", \"tpiA\", \"recF\"]\n",
    "writer = pd.ExcelWriter(f\"{outdir}/priority_expression_report.xlsx\", engine='xlsxwriter')\n",
    "\n",
    "expression_summaries = {}\n",
    "housekeepers_to_annotate = {}\n",
    "numeric_exp = [i for i in derep_df.columns if not i in dereplication_groups['extras']]\n",
    "\n",
    "for bof in bins_of_focus:\n",
    "    bof_df = derep_df.loc[[i for i in derep_df.index if i.startswith(f\"{bof}_\")], numeric_exp].copy()\n",
    "    special_genes = {}\n",
    "    for sgs in special_gene_sets:\n",
    "        for eid_ in gene_sets[sgs]:\n",
    "            if eid_.startswith(bof+\"_\"):\n",
    "                possible_name = \"--\".join(set(pathway_df.loc[pathway_df['identifier'] == eid_, 'gene'].dropna().values))\n",
    "                possible_qid = \"--\".join(set(pathway_df.loc[pathway_df['identifier'] == eid_, 'query_id'].dropna().values))\n",
    "                possible_prod = \"--\".join(set(pathway_df.loc[pathway_df['identifier'] == eid_, 'product'].dropna().values))\n",
    "                special_genes[eid_] = {'bin':bof, 'enzyme':sgs, 'gene_type':'functional', \n",
    "                                       'gene': None, 'title': None, 'hmm':None}\n",
    "                special_genes[eid_].update({'sequence':seq_map[eid_]})\n",
    "                \n",
    "                for field, possib in zip([possible_name, possible_qid, possible_prod], ['gene', 'hmm', 'title']):\n",
    "                    special_genes[eid_].update({possib:field})\n",
    "                if eid_ in set(bof_df.index):\n",
    "                    special_genes[eid_].update(bof_df.loc[eid_, :].to_dict())\n",
    "    \n",
    "    for h in housekeepers:\n",
    "        hbool = (pathway_df.gene == h) & (pathway_df.genome == bof)\n",
    "        if hbool.sum() > 0:\n",
    "            hid_s = set(pathway_df.loc[hbool, 'identifier'])\n",
    "            for eid_ in hid_s:\n",
    "                special_genes[eid_] = {'bin':bof, 'enzyme':h, 'gene_type':'housekeeper', \n",
    "                                       'gene': None, 'title': None, 'hmm':None}\n",
    "                special_genes[eid_].update(bof_df.loc[eid_, :].to_dict())\n",
    "                special_genes[eid_].update({'sequence':seq_map[eid_]})\n",
    "                housekeepers_to_annotate[eid_] = seq_map[eid_]\n",
    "    \n",
    "    topranked = list(bof_df.sum(1).sort_values(ascending=False).index[:50])\n",
    "    needed_cols = ['gene', 'product', 'query_id', 'identifier']\n",
    "    catagg = lambda x: \"--\".join(set(x.dropna()))\n",
    "    highlights_top = pathway_df.loc[pathway_df.identifier.isin(topranked), needed_cols].groupby('identifier').agg(catagg)\n",
    "    highlights_top = highlights_top.drop([i for i in highlights_top.index if i in special_genes])\n",
    "    for eid_ in highlights_top.index:\n",
    "        if not eid_ in special_genes:\n",
    "            special_genes[eid_] = {'bin':bof, 'gene_type':'ranked', 'hmm':None, 'enzyme': None,\n",
    "                                   'title':highlights_top.loc[eid_, 'product'], \n",
    "                                   'gene':highlights_top.loc[eid_, 'gene']}\n",
    "            special_genes[eid_].update(bof_df.loc[eid_, :].to_dict())\n",
    "            special_genes[eid_].update({'sequence':seq_map[eid_]})\n",
    "        else:\n",
    "            special_genes[eid_]['gene_type'] = special_genes[eid_]['gene_type'] + \"+ranked\"\n",
    "    \n",
    "    expression_summary = pd.DataFrame(special_genes).T\n",
    "    bad_cols = np.array(numeric_exp)[expression_summary.loc[expression_summary.enzyme.isin(housekeepers), numeric_exp].sum() == 0]\n",
    "    to_drop = highlights_top.index[(expression_summary.loc[list(highlights_top.index), bad_cols] > 0).sum(1) != 0]\n",
    "    expression_summary.drop(to_drop, axis=0, inplace=True)\n",
    "    expression_summary = expression_summary.loc[:, [i for i in expression_summary.columns if i != 'sequence']+['sequence']]\n",
    "    expression_summary.to_excel(writer, sheet_name=f'{bof}')\n",
    "    expression_summaries[bof] = expression_summary.copy()\n",
    "\n",
    "writer.save()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "nuclear-projector",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "a = derep_df.loc[gene_sets['nirABD'] & set(derep_df.index) , :]\n",
    "b = derep_df.loc[gene_sets['nirK/nirS+norBC'] & set(derep_df.index), :]\n",
    "print(a.mean())\n",
    "print(b.mean())\n",
    "print(a.shape, b.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "instrumental-projector",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "instructional-sociology",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame(index=bins_of_focus, columns=list(special_gene_sets | set(gene_cats[1]))).T.fillna(0)\n",
    "print(df.shape)\n",
    "print(set(df.index) - set(gene_cats[1]))\n",
    "print(set(gene_cats[1]) - set(df.index))\n",
    "\n",
    "for b in df.columns:\n",
    "    print(b, report_df.loc[b, 'taxonomic leaf name'])\n",
    "    exp_sum_i = expression_summaries[b]\n",
    "    hslice = exp_sum_i.loc[exp_sum_i.gene_type == 'housekeeper', numeric_exp].mean()\n",
    "    \n",
    "    for gene in df.index:\n",
    "        if gene in housekeepers:\n",
    "            hbool = (pathway_df.gene == gene) & (pathway_df.genome == b)\n",
    "            if hbool.sum() > 0:\n",
    "                df.loc[gene, b] += 1\n",
    "        elif gene in set(annot_mat.columns):\n",
    "            if gene in set(annot_mat.columns[annot_mat.loc[b, :]]):\n",
    "                df.loc[gene, b] += 1\n",
    "        else:\n",
    "            raise ValueError(\"Missing gene from annot mat / housekeepers\")\n",
    "    \n",
    "        if gene in set(exp_sum_i.enzyme):\n",
    "            gslice = exp_sum_i.loc[exp_sum_i.enzyme == gene, numeric_exp]\n",
    "            df.loc[gene, b] += int(((gslice.mean() / hslice.mean()) > 0).sum() > 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fourth-premium",
   "metadata": {},
   "outputs": [],
   "source": [
    "figdir = f\"{os.path.dirname(outdir)}/Figures\"\n",
    "df = df.loc[[i for i in gene_cats[1] if i in set(df.index)], :]\n",
    "gene_names_and_cats = {i:\"{} ({})\".format(i, j.lower()) for i, j in gene_cats.set_index(1)[0].to_dict().items()}\n",
    "bin_names_and_taxa = {i:\"{} ({})\".format(i, report_df.loc[i, 'taxonomic leaf name']) for i in df.columns}\n",
    "dfp = df.rename(index=gene_names_and_cats, columns=bin_names_and_taxa)\n",
    "\n",
    "# plot\n",
    "\n",
    "myColors = ((1.0, 1.0, 1.0, 0.0), \n",
    "            (0.4, 0.4, 0.4, 0.4), \n",
    "            (0.0, 0.0, 0.0, 1.0))\n",
    "cmap = LinearSegmentedColormap.from_list('Custom', myColors, len(myColors))\n",
    "\n",
    "sns.set(font=\"Calibri\", font_scale=0.74)\n",
    "plt.close('all')\n",
    "fig = plt.figure(constrained_layout=True, figsize=(6*1.2, 8*1.2), dpi=300, num=1)\n",
    "gs = fig.add_gridspec(8, 6)\n",
    "ax = fig.add_subplot(gs[:-1, :])\n",
    "ax2 = fig.add_subplot(gs[-1, 0])\n",
    "#ax3 = fig.add_subplot(gs[-1, 0])\n",
    "cm_fig = sns.heatmap(dfp, cbar_ax=ax2, ax=ax, cmap=cmap, linewidths=.2, cbar_kws={'drawedges':True})\n",
    "ax.set_xticklabels(ax.get_xticklabels(),rotation=30, ha=\"right\")\n",
    "ax2.set_yticklabels(['Not Observed', 'Annotation Observed', '>1x expression observed'])\n",
    "ax2.yaxis.set_ticks([ 0.3, 0.9, 1.6])\n",
    "fig.savefig(f'{figdir}/gene_repertoire.png', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "related-terrorist",
   "metadata": {},
   "outputs": [],
   "source": [
    "transcripts_detected = (((derep_df.loc[:, numeric_exp] > 0).sum().sort_index() / len(derep_df))*100)\n",
    "plt.figure(figsize=(10, 8))\n",
    "\n",
    "n, bins, patches = plt.hist(x=transcripts_detected, bins=10, color='#0504aa', alpha=0.7, rwidth=0.85)\n",
    "\n",
    "plt.show()\n",
    "print((transcripts_detected < 3).sum(), len(transcripts_detected))\n",
    "transcripts_detected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "pending-burns",
   "metadata": {},
   "outputs": [],
   "source": [
    "dsrdudes = ['CB33.bin.149', 'CBrest.bin.263', 'CB33.bin.191', 'CB33.bin.233', 'CB33.bin.196', \n",
    "            'CBrest.bin.358', 'CB33.bin.236', 'CBrest.bin.394', 'CB33.bin.235', 'CBrest.bin.368',\n",
    "            'CB33.bin.202', 'CB33.bin.216', 'CB33.bin.174', 'CB33.bin.186', 'CBrest.bin.322', 'CB33.bin.118',\n",
    "            'CB33.bin.222', 'CB33.bin.175']\n",
    "qualtax_df.loc[dsrdudes, ['Phylum', 'Class', 'Order', 'Family', 'Genus']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "hourly-closure",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "engaged-server",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "cb33_cols = [i for i in set(copy_num_df.columns) & set(cb33_abnd2.columns) if not i.startswith('Zymo') and not i.startswith('SERC')]\n",
    "cb33_tuples = [i.replace(\"CB33_\", \"\").replace(\"June\", \"6\").replace(\"Aug\", \"8\").replace(\"July\", \"7\").replace(\"_2017\", \"\") for i in cb33_cols]\n",
    "cb33_tuples = [i.replace(\"_R2\", \"\").replace(\"R2\", \"\").replace(\"_R7\", \"\").replace(\"_R1\", \"\").split(\"_\") for i in cb33_tuples]\n",
    "cb33_tuples = [[i[0], i[-1]] for i in cb33_tuples]\n",
    "cb33_df = pd.DataFrame(index=cb33_cols, data=cb33_tuples)\n",
    "cb33_df[2] = cb33_df[1].apply(lambda x: int(x.replace(\"m\", \"\")))\n",
    "cb33_df = cb33_df.sort_values(by=[0, 2])\n",
    "cb33_df['label'] = cb33_df[0] + \"-33-\" + cb33_df[1]\n",
    "\n",
    "cbRest_df = pd.DataFrame(index=cbRest_cols, data=cbRest_tuples)\n",
    "\n",
    "cbRest_cols = [i for i in set(copy_num_df.columns) - set(cb33_abnd2.columns) if not 'Bay_Bridge' in i and not 'Negative' in i]\n",
    "cbRest_tuples = [i.split(\"_\") for i in cbRest_cols]\n",
    "cbRest_df = pd.DataFrame(index=cbRest_cols, data=cbRest_tuples)\n",
    "cbRest_df.loc[cbRest_df[0] == '4102017', 0] = '41017'\n",
    "cbRest_df.loc[cbRest_df[0] == '6617', 0] = '60617'\n",
    "cbRest_df.loc[cbRest_df[0] == '6517', 0] = '60517'\n",
    "cbRest_df.loc[cbRest_df[1] == 'CB53-1', 1] = 'CB53'\n",
    "cbRest_df[3] = cbRest_df[0].apply(lambda x: x[0])\n",
    "cbRest_df[4] = cbRest_df[1].apply(lambda x: x.replace(\"CB\", \"\").replace('C', \"\"))\n",
    "cbRest_df = cbRest_df.sort_values(by=[2,3,4])\n",
    "cbRest_df['label'] = cbRest_df[3] + \"-\" + cbRest_df[4] + \"-\" + cbRest_df[2]\n",
    "\n",
    "cb33copies = copy_num_df[list(cb33_df.index)]\n",
    "cbRestcopies = copy_num_df[list(cbRest_df.index)]\n",
    "\n",
    "exp_cols = [i for i in salmon_df.columns if not i == 'merger_col']\n",
    "exp_tuples = [i.split(\"_\") for i in exp_cols]\n",
    "exp_df = pd.DataFrame(index=exp_cols, data=exp_tuples)\n",
    "exp_df[4] = exp_df[0].apply(lambda x: x.split(\"/\")[0])\n",
    "exp_df['label'] = exp_df[4] + \"-43-\" + exp_df[1]\n",
    "exp_df = exp_df.sort_values(by=[1, 4])\n",
    "\n",
    "import matplotlib.gridspec as gridspec\n",
    "\n",
    "plt.close('all'); plt.clf();\n",
    "fig2 = plt.figure(constrained_layout=True, figsize=(11, 8), dpi=300, num=2)\n",
    "gs2 = gridspec.GridSpec(ncols=10, nrows=11, figure=fig2)\n",
    "all_axes = {}\n",
    "for b_i, bof in enumerate(bins_of_focus):\n",
    "    spec_ax = {}\n",
    "    spec_ax['name'] = fig2.add_subplot(gs2[b_i, 0], facecolor=\"white\", frameon=True)\n",
    "    _ = spec_ax['name'].set_xlim(-1, 1.)\n",
    "    _ = spec_ax['name'].set_ylim(-1, 1.)\n",
    "    _ = spec_ax['name'].axes.get_xaxis().set_ticks([])\n",
    "    _ = spec_ax['name'].axes.get_yaxis().set_ticks([])\n",
    "    _ = spec_ax['name'].text(0, .5, bof, ha=\"center\", va=\"center\")\n",
    "    _ = spec_ax['name'].text(0, -.5, report_df.loc[bof, 'taxonomic leaf name'], ha=\"center\", va=\"center\")\n",
    "    ##################################################\n",
    "    spec_ax['cb33'] = fig2.add_subplot(gs2[b_i, 1:4], facecolor=\"white\", frameon=True)\n",
    "    cleaned_2 = set(pathway_df.loc[pathway_df['genome'] == bof, 'GeneID'].dropna())\n",
    "    cleaned_4 = cleaned_2 & set(copy_num_df.index)\n",
    "    cb33_slice = cb33copies.loc[cleaned_4, :].apply(np.median).values\n",
    "    cb33_upper = cb33copies.loc[cleaned_4, :].apply(lambda x: np.percentile(x, 75)).values - cb33_slice\n",
    "    cb33_lower = cb33_slice - cb33copies.loc[cleaned_4, :].apply(lambda x: np.percentile(x, 25)).values\n",
    "    cb33_yerr  = np.vstack((cb33_upper, cb33_lower))\n",
    "    _ = spec_ax['cb33'].errorbar(np.arange(len(cb33_df)), cb33_slice, yerr=cb33_yerr, \n",
    "            color='k', marker='.', capsize=1, capthick=1, ecolor='black')\n",
    "    # _ = spec_ax['cb33'].plot(np.arange(len(cb33_df)), cb33_slice, color='k')\n",
    "    if b_i == (len(bins_of_focus) - 1):\n",
    "        _ = spec_ax['cb33'].set_xticks(np.arange(len(cb33_df)))\n",
    "        _ = spec_ax['cb33'].set_xticklabels(list(cb33_df['label']), rotation=90, ha=\"center\")\n",
    "    else:\n",
    "        _ = spec_ax['cb33'].axes.get_xaxis().set_ticks([])\n",
    "    ##################################################\n",
    "    spec_ax['cbrest'] = fig2.add_subplot(gs2[b_i, 4:8], facecolor=\"white\", frameon=True)\n",
    "    cbRest_slice = cbRestcopies.loc[cleaned_4, :].apply(np.median).values\n",
    "    _ = spec_ax['cbrest'].plot(np.arange(len(cbRest_slice)), cbRest_slice, color='k')\n",
    "    if b_i == (len(bins_of_focus) - 1):\n",
    "        _ = spec_ax['cbrest'].set_xticks(np.arange(len(cbRest_slice)))\n",
    "        _ = spec_ax['cbrest'].set_xticklabels(list(cbRest_df['label']), rotation=90, ha=\"center\")\n",
    "    else:\n",
    "        _ = spec_ax['cbrest'].axes.get_xaxis().set_ticks([])\n",
    "    ##################################################\n",
    "    spec_ax['mrna'] = fig2.add_subplot(gs2[b_i, 8:10], facecolor=\"white\", frameon=True)\n",
    "    exp_slice = salmon_dfs[bof].loc[:, exp_df.index].apply(np.median).values\n",
    "    _ = spec_ax['mrna'].plot(np.arange(len(exp_df)), exp_slice, color='k')\n",
    "    if b_i == (len(bins_of_focus) - 1):\n",
    "        _ = spec_ax['mrna'].set_xticks(np.arange(len(exp_slice)))\n",
    "        _ = spec_ax['mrna'].set_xticklabels(list(exp_df['label']), rotation=90, ha=\"center\")\n",
    "    else:\n",
    "        _ = spec_ax['mrna'].axes.get_xaxis().set_ticks([])\n",
    "    ##################################################\n",
    "    all_axes[bof] = spec_ax\n",
    "\n",
    "all_axes['annotations'] = fig2.add_subplot(gs2[-1, :], facecolor=\"white\", frameon=True)\n",
    "_ = all_axes['annotations'].axes.get_xaxis().set_ticks([])\n",
    "_ = all_axes['annotations'].axes.get_yaxis().set_ticks([])\n",
    "fig2.savefig('/Volumes/KeithSSD/SulFox/figures/bin_location.png', dpi=300)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
